{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "32f4fa11-1ac9-4a54-bab3-277e27f115bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import admix\n",
    "import numpy as np\n",
    "import glob\n",
    "from os.path import join\n",
    "import os\n",
    "import itertools\n",
    "import submitit\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "67f953d5-5e0b-4069-b9af-1067bc6df373",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPP_TABLE_URL = \"https://www.dropbox.com/s/jck2mhjby2ur55j/supp-tables.xlsx?dl=1\"\n",
    "\n",
    "df_trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"ukb-trait-info\")\n",
    "ukb_trait_list = df_trait_info[df_trait_info[\"in-analysis\"] == \"T\"].id.values\n",
    "\n",
    "df_trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"page-trait-info\")\n",
    "page_trait_list = df_trait_info.trait.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "81b3fecf-49c0-493f-82e0-bf39c3c3c0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "PAGE_GENO_DIR = \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01-dataset/out/aframr/imputed/\"\n",
    "PAGE_PHENO_DIR = (\n",
    "    \"/u/project/pasaniuc/kangchen/2021-admix-corr/experiments/\"\n",
    "    \"03-page-genome-wide-profile-likelihood-new/out/pheno\"\n",
    ")\n",
    "\n",
    "UKB_GENO_DIR = \"/u/project/sgss/UKBB/UKB-ADMIXED/01-dataset/out/PLINK2/imputed\"\n",
    "UKB_PHENO_DIR = \"/u/project/sgss/UKBB/UKB-ADMIXED/02-genet-cor/out/pheno/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d761a301-6d57-4778-94b7-4401d06cbdf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-06-03 12:51.53 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.53 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.53 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.53 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.53 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.53 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.53 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.53 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.54 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.54 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.55 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.55 [info     ] Added dummy variables: study_WHI\n",
      "2022-06-03 12:51.56 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.56 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.56 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.56 [info     ] Added dummy variables: study_MEC,study_WHI\n",
      "2022-06-03 12:51.56 [info     ] Detected categorical columns: study\n",
      "2022-06-03 12:51.56 [info     ] Added dummy variables: study_WHI\n"
     ]
    }
   ],
   "source": [
    "for study in [\"page\", \"ukb\"]:\n",
    "    if study == \"page\":\n",
    "        trait_list = page_trait_list\n",
    "    else:\n",
    "        trait_list = ukb_trait_list\n",
    "\n",
    "    pheno_dir = UKB_PHENO_DIR if study == \"ukb\" else PAGE_PHENO_DIR\n",
    "    for trait in trait_list:\n",
    "        df_trait = pd.read_csv(\n",
    "            os.path.join(pheno_dir, f\"{trait}-sample10pc.tsv\"),\n",
    "            sep=\"\\t\",\n",
    "            index_col=0,\n",
    "            low_memory=False,\n",
    "        )\n",
    "        pheno_col = df_trait.columns[0]\n",
    "        covar_cols = df_trait.columns[1:]\n",
    "        df_trait[pheno_col] = df_trait[pheno_col] - df_trait[pheno_col].mean()\n",
    "        df_trait = df_trait.rename(columns={pheno_col: \"pheno\"})\n",
    "        df_trait = admix.data.convert_dummy(df_trait)\n",
    "        df_trait.loc[:, df_trait.columns] = admix.data.impute_with_mean(\n",
    "            df_trait.values, axis=0\n",
    "        )\n",
    "        df_trait.index.name = \"#IID\"\n",
    "        df_trait.iloc[:, [0]].to_csv(\n",
    "            f\"out/processed-pheno/{study}/{trait}.pheno.tsv\", sep=\"\\t\", na_rep=\"NA\"\n",
    "        )\n",
    "        df_trait.iloc[:, 1:].to_csv(\n",
    "            f\"out/processed-pheno/{study}/{trait}.covar.tsv\", sep=\"\\t\", na_rep=\"NA\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fc2dbec4-a530-4ed3-a48a-8e4e280e2d9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate_het(study, chrom):\n",
    "\n",
    "    geno_dir = UKB_GENO_DIR if study == \"ukb\" else PAGE_GENO_DIR\n",
    "    pfile = join(geno_dir, f\"chr{chrom}\")\n",
    "    bfile = join(geno_dir, f\"PLINK1/chr{chrom}\")\n",
    "    if study == \"page\":\n",
    "        trait_list = page_trait_list\n",
    "    else:\n",
    "        trait_list = ukb_trait_list\n",
    "\n",
    "    for trait in trait_list:\n",
    "        pheno_file = f\"out/processed-pheno/{study}/{trait}.pheno.tsv\"\n",
    "        covar_file = f\"out/processed-pheno/{study}/{trait}.covar.tsv\"\n",
    "        out_prefix = f\"out/gwas/{study}/{trait}.chr{chrom}\"\n",
    "\n",
    "        # step 1: association testing\n",
    "        cmds = [\n",
    "            \"plink2\",\n",
    "            f\"--bfile {bfile}\",\n",
    "            f\"--pheno iid-only {pheno_file}\",\n",
    "            f\"--covar iid-only {covar_file}\",\n",
    "            \"--quantile-normalize --glm hide-covar omit-ref --vif 100\",\n",
    "            \"--memory 20000\",\n",
    "            f\"--out {out_prefix}\",\n",
    "        ]\n",
    "        subprocess.check_call(\" \".join(cmds), shell=True)\n",
    "        subprocess.check_call(\n",
    "            f\"sed '1 s/ID/SNP/' {out_prefix}.pheno.glm.linear > {out_prefix}.assoc\",\n",
    "            shell=True,\n",
    "        )\n",
    "\n",
    "        # step 2: clumping\n",
    "        cmds = [\n",
    "            \"plink\",\n",
    "            f\"--bfile {bfile}\",\n",
    "            f\"--clump {out_prefix}.assoc\",\n",
    "            f\"--clump-p1 5e-8 --clump-p2 1e-4 --clump-r2 0.1 --clump-kb 10000 --memory 20000\",\n",
    "            f\"--out {out_prefix}\",\n",
    "        ]\n",
    "        subprocess.check_call(\" \".join(cmds), shell=True)\n",
    "        if not os.path.exists(f\"{out_prefix}.clumped\"):\n",
    "            continue\n",
    "\n",
    "        subprocess.check_call(\n",
    "            f\"awk '(NF > 0) && (NR > 1) {{print $3 }}' {out_prefix}.clumped > {out_prefix}.clumped_snp_list\",\n",
    "            shell=True,\n",
    "        )\n",
    "\n",
    "        # step 3: heterogeneity\n",
    "        cmds = [\n",
    "            \"admix assoc\",\n",
    "            f\"--pfile {pfile}\",\n",
    "            f\"--pheno {pheno_file}\",\n",
    "            f\"--covar {covar_file}\",\n",
    "            f\"--out {out_prefix}\",\n",
    "            \"--method HET,ADM,ASE,ATT,TRACTOR\",\n",
    "            \"--family quant\",\n",
    "            \"--quantile-normalize True\",\n",
    "            f\"--snp-list {out_prefix}.clumped_snp_list\",\n",
    "        ]\n",
    "        subprocess.check_call(\n",
    "            \" \".join(cmds),\n",
    "            shell=True,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4b5cbf33-e9ef-4166-93af-0e8de80cc045",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_params = pd.DataFrame(\n",
    "    [param for param in itertools.product([\"ukb\", \"page\"], np.arange(1, 23))],\n",
    "    columns=[\"study\", \"chrom\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b55c539a-b55b-4210-bd02-4649917749ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=180,\n",
    "    memory_g=24,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PATH=~/project-pasaniuc/software/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    evaluate_het,\n",
    "    df_params.study,\n",
    "    df_params.chrom,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
