{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b9a3a36-9d68-463b-b58a-cb58a5932b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import admix\n",
    "import dask\n",
    "import dask.array as da\n",
    "from tqdm import tqdm\n",
    "import statsmodels.api as sm\n",
    "import matplotlib.pyplot as plt\n",
    "from admix.data import quantile_normalize\n",
    "import submitit\n",
    "import dapgen\n",
    "import os\n",
    "import admix_genet_cor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b713c7e7-1fc6-4f7e-9f37-9903688f66e7",
   "metadata": {},
   "source": [
    "# Process GWAS hits from PAGE / UKB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d32e6ce5-efff-4813-85f6-2430f675ae06",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPP_TABLE_URL = \"https://www.dropbox.com/s/jck2mhjby2ur55j/supp-tables.xlsx?dl=1\"\n",
    "\n",
    "df_trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"ukb-trait-info\")\n",
    "ukb_trait_list = df_trait_info[df_trait_info[\"in-analysis\"] == \"T\"].id.values\n",
    "\n",
    "df_trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"page-trait-info\")\n",
    "page_trait_list = df_trait_info.trait.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3c4ff7f5-0ab6-42a7-8e02-16623a3ee405",
   "metadata": {},
   "outputs": [],
   "source": [
    "PAGE_GENO_DIR = \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01-dataset/out/aframr/imputed/\"\n",
    "PAGE_PHENO_DIR = \"/u/project/pasaniuc/kangchen/2021-admix-corr/experiments/03-page-genome-wide-profile-likelihood/out/pheno\"\n",
    "\n",
    "UKB_GENO_DIR = \"/u/project/sgss/UKBB/UKB-ADMIXED/01-dataset/out/PLINK2/imputed\"\n",
    "UKB_PHENO_DIR = \"/u/project/sgss/UKBB/UKB-ADMIXED/02-genet-cor/out/pheno/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "96447c4c-a6af-471a-8ed7-c7a05a6295a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_locus_hetero(study, trait, out):\n",
    "\n",
    "    # load clumped variants\n",
    "    clump_file = f\"out/gwas-assoc/{study}-{trait}.clumped\"\n",
    "    if sum(1 for line in open(clump_file)) == 1:\n",
    "        with open(out, \"w\") as f:\n",
    "            f.writelines(\"# No clumped region\")\n",
    "        return\n",
    "\n",
    "    df_assoc = (\n",
    "        pd.read_csv(f\"out/gwas-assoc/{study}-{trait}.clumped\", sep=\"\\t\")[\n",
    "            [\"CHR\", \"SNP\", \"BP\", \"P\"]\n",
    "        ]\n",
    "        .rename(columns={\"CHR\": \"CHROM\", \"SNP\": \"SNP\", \"BP\": \"POS\", \"P\": \"PLINK_P\"})\n",
    "        .set_index(\"SNP\")\n",
    "    )\n",
    "\n",
    "    # load both phenotype and covariates\n",
    "    geno_dir = UKB_GENO_DIR if study == \"ukb\" else PAGE_GENO_DIR\n",
    "    pheno_dir = UKB_PHENO_DIR if study == \"ukb\" else PAGE_PHENO_DIR\n",
    "\n",
    "    pheno_path = os.path.join(pheno_dir, f\"{trait}-sample10pc.tsv\")\n",
    "    df_sample_info = pd.read_csv(pheno_path, sep=\"\\t\", index_col=0, low_memory=False)\n",
    "\n",
    "    pheno_col = df_sample_info.columns[0]\n",
    "    covar_cols = df_sample_info.columns[1:]\n",
    "\n",
    "    # format phenotype and covariates\n",
    "    df_trait = pd.read_csv(pheno_path, sep=\"\\t\", index_col=0)\n",
    "    df_trait.index = df_trait.index.astype(str)\n",
    "\n",
    "    dset = admix.io.read_dataset(\n",
    "        os.path.join(geno_dir, \"chr1\"),\n",
    "        n_anc=2,\n",
    "    )\n",
    "    for col in dset._indiv:\n",
    "        dset._indiv.drop(columns=col, inplace=True)\n",
    "\n",
    "    # subset for individuals with non-nan value in df_trait\n",
    "    dset = dset[:, dset.indiv.index.isin(df_trait.index)]\n",
    "    dset.append_indiv_info(df_trait)\n",
    "    pheno_col = df_trait.columns[[0]]\n",
    "    covar_cols = df_trait.columns[1:]\n",
    "\n",
    "    df_pheno = dset.indiv[pheno_col].copy()\n",
    "    df_covar = dset.indiv[covar_cols].copy()\n",
    "    df_covar = admix.data.convert_dummy(df_covar)\n",
    "\n",
    "    for col in df_pheno.columns:\n",
    "        df_pheno[col] = admix.data.quantile_normalize(df_pheno[col])\n",
    "\n",
    "    for col in df_covar.columns:\n",
    "        df_covar[col] = admix.data.quantile_normalize(df_covar[col])\n",
    "\n",
    "    # fill na with column mean\n",
    "    df_covar.fillna(df_covar.mean(), inplace=True)\n",
    "\n",
    "    for chrom in range(1, 23):\n",
    "        # subset df_assoc\n",
    "        df_assoc_chrom = df_assoc[df_assoc.CHROM == chrom]\n",
    "        if len(df_assoc_chrom) == 0:\n",
    "            continue\n",
    "        # subset dset\n",
    "        dset = admix.io.read_dataset(\n",
    "            os.path.join(geno_dir, f\"chr{chrom}\"),\n",
    "            n_anc=2,\n",
    "        )\n",
    "        dset = dset[df_assoc_chrom.index.values, df_pheno.index.values]\n",
    "        dset.persist()\n",
    "\n",
    "        apa = dset.allele_per_anc()\n",
    "        af = dset.af_per_anc()\n",
    "\n",
    "        for snp_i in tqdm(range(dset.n_snp)):\n",
    "            p_het, model_het = admix_genet_cor.test_snp_het(\n",
    "                apa[snp_i, :, :], df_pheno.values, df_covar.values\n",
    "            )\n",
    "            p_assoc, model_assoc = admix_genet_cor.test_snp_assoc(\n",
    "                apa[snp_i, :, :], df_pheno.values, df_covar.values\n",
    "            )\n",
    "            snp_idx = dset.snp.index[snp_i]\n",
    "            df_assoc.loc[snp_idx, [\"EUR_af\", \"AFR_af\"]] = (\n",
    "                af[snp_i, 0],\n",
    "                af[snp_i, 1],\n",
    "            )\n",
    "\n",
    "            df_assoc.loc[\n",
    "                snp_idx,\n",
    "                [\n",
    "                    \"assoc_pval\",\n",
    "                    \"HET_pval\",\n",
    "                    \"EUR_beta\",\n",
    "                    \"AFR_beta\",\n",
    "                    \"EUR_beta_stderr\",\n",
    "                    \"AFR_beta_stderr\",\n",
    "                ],\n",
    "            ] = [\n",
    "                p_assoc,\n",
    "                p_het,\n",
    "                model_het.params[1],\n",
    "                model_het.params[2],\n",
    "                model_het.bse[1],\n",
    "                model_het.bse[2],\n",
    "            ]\n",
    "    df_assoc[[\"CHROM\", \"POS\"]] = df_assoc[[\"CHROM\", \"POS\"]].astype(int)\n",
    "    df_assoc.sort_values([\"CHROM\", \"POS\"]).to_csv(out, float_format=\"%.8g\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f1a060ec-c382-4e2c-8fcc-e53fc98e73dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_params = {\"study\": [], \"trait\": []}\n",
    "df_params[\"trait\"].extend(ukb_trait_list)\n",
    "df_params[\"study\"].extend([\"ukb\"] * len(ukb_trait_list))\n",
    "df_params[\"trait\"].extend(page_trait_list)\n",
    "df_params[\"study\"].extend([\"page\"] * len(page_trait_list))\n",
    "df_params = pd.DataFrame(df_params)\n",
    "df_params[\"out\"] = df_params.apply(\n",
    "    lambda x: f\"out/gwas-het/{x.study}-{x.trait}.csv\", axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "019fd055-ed3f-434f-9613-e4d0b4ccb2f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=40,\n",
    "    memory_g=16,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_locus_hetero,\n",
    "    df_params.study,\n",
    "    df_params.trait,\n",
    "    df_params.out,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
