{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b41ad593-351f-43ab-bf83-1d8b572293fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext lab_black\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from admix.data import quantile_normalize\n",
    "import seaborn as sns\n",
    "import admix\n",
    "import admix_genet_cor\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "52ef9b90-88cd-4149-bcd7-5043dbb9cae7",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPP_TABLE_URL = \"https://www.dropbox.com/s/jck2mhjby2ur55j/supp-tables.xlsx?dl=1\"\n",
    "\n",
    "df_trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"ukb-trait-info\")\n",
    "ukb_trait_list = df_trait_info[df_trait_info[\"in-analysis\"] == \"T\"].id.values\n",
    "\n",
    "df_trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"page-trait-info\")\n",
    "page_trait_list = df_trait_info.trait.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "08604826-54ad-4a8c-8972-5ed937ba6ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "PAGE_GENO_DIR = \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01-dataset/out/aframr/imputed/\"\n",
    "PAGE_PHENO_DIR = \"/u/project/pasaniuc/kangchen/2021-admix-corr/experiments/03-page-genome-wide-profile-likelihood/out/pheno\"\n",
    "\n",
    "UKB_GENO_DIR = \"/u/project/sgss/UKBB/UKB-ADMIXED/01-dataset/out/PLINK2/imputed\"\n",
    "UKB_PHENO_DIR = \"/u/project/sgss/UKBB/UKB-ADMIXED/02-genet-cor/out/pheno/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9b795887-9b6e-4644-84b6-06d948ab579c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_params = {\"study\": [], \"trait\": []}\n",
    "df_params[\"trait\"].extend(ukb_trait_list)\n",
    "df_params[\"study\"].extend([\"ukb\"] * len(ukb_trait_list))\n",
    "df_params[\"trait\"].extend(page_trait_list)\n",
    "df_params[\"study\"].extend([\"page\"] * len(page_trait_list))\n",
    "df_params = pd.DataFrame(df_params)\n",
    "df_params[\"out\"] = df_params.apply(\n",
    "    lambda x: f\"out/gwas-het/{x.study}-{x.trait}.csv\", axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "94bbbbf2-e56e-4eec-b710-d553a0c6343f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "out/gwas-het/ukb-height.csv is empty\n",
      "out/gwas-het/ukb-log_BMI.csv is empty\n",
      "out/gwas-het/ukb-log_heel_BMD.csv is empty\n",
      "out/gwas-het/ukb-log_leukocyte.csv is empty\n",
      "out/gwas-het/ukb-log_lymphocyte.csv is empty\n",
      "out/gwas-het/ukb-log_platelet.csv is empty\n",
      "out/gwas-het/ukb-neuroticism.csv is empty\n",
      "out/gwas-het/ukb-years_of_edu.csv is empty\n",
      "out/gwas-het/page-a1c.csv is empty\n",
      "out/gwas-het/page-insulin.csv is empty\n",
      "out/gwas-het/page-qt_interval.csv is empty\n",
      "out/gwas-het/page-qrs_interval.csv is empty\n",
      "out/gwas-het/page-systolic_bp.csv is empty\n",
      "out/gwas-het/page-diastolic_bp.csv is empty\n",
      "out/gwas-het/page-hypertension.csv is empty\n",
      "out/gwas-het/page-waist_hip_ratio.csv is empty\n"
     ]
    }
   ],
   "source": [
    "df_assoc = []\n",
    "for i, param in df_params.iterrows():\n",
    "    df_tmp = pd.read_csv(param.out)\n",
    "    if len(df_tmp) == 0:\n",
    "        print(param.out, \"is empty\")\n",
    "        continue\n",
    "    df_tmp = pd.DataFrame({\"study\": [param.study], \"trait\": [param.trait]}).merge(\n",
    "        df_tmp, how=\"cross\"\n",
    "    )\n",
    "    df_assoc.append(df_tmp)\n",
    "df_assoc = pd.concat(df_assoc).sort_values([\"trait\", \"CHROM\", \"POS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "76c4acf9-c7f7-42ed-bcb2-58155e4225d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_assoc = df_assoc[\n",
    "    df_assoc.EUR_af.between(0.005, 0.995) & df_assoc.AFR_af.between(0.005, 0.995)\n",
    "].reset_index(drop=True)\n",
    "df_assoc.loc[:, \"expected_HET_pval\"] = stats.norm.sf(\n",
    "    quantile_normalize(-df_assoc.HET_pval)\n",
    ")\n",
    "df_assoc.loc[:, \"bonferroni_significant\"] = df_assoc[\"HET_pval\"] < 0.05 / len(df_assoc)\n",
    "\n",
    "df_assoc = df_assoc[df_assoc[\"assoc_pval\"] < 5e-8]\n",
    "# df_assoc = df_assoc[~df_assoc.trait.isin([\"MCH\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "098d8739-f60f-4785-b0c8-69c49ec85f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "def report_assoc(df_assoc):\n",
    "\n",
    "    bootstrap_slope_list = []\n",
    "    pearsonr_list = []\n",
    "    for _ in range(1000):\n",
    "        df_tmp = df_assoc.sample(len(df_assoc), replace=True)\n",
    "        slope = admix_genet_cor.locus.deming_regression(\n",
    "            x=df_tmp[\"EUR_beta\"],\n",
    "            y=df_tmp[\"AFR_beta\"],\n",
    "            sx=df_tmp[\"EUR_beta_stderr\"],\n",
    "            sy=df_tmp[\"AFR_beta_stderr\"],\n",
    "        )[0]\n",
    "        bootstrap_slope_list.append(slope)\n",
    "        pearsonr = stats.pearsonr(df_tmp.EUR_beta, df_tmp.AFR_beta)[0]\n",
    "        pearsonr_list.append(pearsonr)\n",
    "\n",
    "    slope = admix_genet_cor.locus.deming_regression(\n",
    "        x=df_assoc[\"EUR_beta\"],\n",
    "        y=df_assoc[\"AFR_beta\"],\n",
    "        sx=df_assoc[\"EUR_beta_stderr\"],\n",
    "        sy=df_assoc[\"AFR_beta_stderr\"],\n",
    "    )[0]\n",
    "    lgc, lgc_ci = admix.data.lambda_gc(df_assoc.HET_pval, bootstrap_ci=True)\n",
    "\n",
    "    print(f\"{len(df_assoc)} SNPs\")\n",
    "    print(\n",
    "        f\"Pearson's r = {stats.pearsonr(df_assoc.EUR_beta, df_assoc.AFR_beta)[0]:.2g}, SE={np.std(pearsonr_list):.2f}\"\n",
    "    )\n",
    "    print(\n",
    "        f\"Deming regression slope: {slope:.2f}. Bootstrap mean (se): {np.mean(bootstrap_slope_list):.2f} ({np.std(bootstrap_slope_list):.2f})\"\n",
    "    )\n",
    "    print(f\"lambda gc = {lgc:.2f}, ci=[{lgc_ci[0]:.2f}, {lgc_ci[1]:.2f}]\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "415a3a01-d5f6-4517-89a3-8251ba803366",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>trait</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>PLINK_P</th>\n",
       "      <th>EUR_af</th>\n",
       "      <th>AFR_af</th>\n",
       "      <th>assoc_pval</th>\n",
       "      <th>HET_pval</th>\n",
       "      <th>EUR_beta</th>\n",
       "      <th>AFR_beta</th>\n",
       "      <th>EUR_beta_stderr</th>\n",
       "      <th>AFR_beta_stderr</th>\n",
       "      <th>expected_HET_pval</th>\n",
       "      <th>bonferroni_significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:140238:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>140238</td>\n",
       "      <td>6.540000e-10</td>\n",
       "      <td>0.428832</td>\n",
       "      <td>0.772786</td>\n",
       "      <td>6.542885e-10</td>\n",
       "      <td>1.854274e-24</td>\n",
       "      <td>0.074391</td>\n",
       "      <td>-0.277259</td>\n",
       "      <td>0.031295</td>\n",
       "      <td>0.026369</td>\n",
       "      <td>0.003906</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:207212:A:C</td>\n",
       "      <td>16</td>\n",
       "      <td>207212</td>\n",
       "      <td>2.840000e-09</td>\n",
       "      <td>0.490268</td>\n",
       "      <td>0.844401</td>\n",
       "      <td>2.843981e-09</td>\n",
       "      <td>4.842501e-24</td>\n",
       "      <td>0.049764</td>\n",
       "      <td>-0.289335</td>\n",
       "      <td>0.031103</td>\n",
       "      <td>0.028097</td>\n",
       "      <td>0.011719</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:291988:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>291988</td>\n",
       "      <td>5.680000e-14</td>\n",
       "      <td>0.177670</td>\n",
       "      <td>0.247776</td>\n",
       "      <td>5.677463e-14</td>\n",
       "      <td>1.097601e-21</td>\n",
       "      <td>0.124169</td>\n",
       "      <td>-0.371366</td>\n",
       "      <td>0.042412</td>\n",
       "      <td>0.031606</td>\n",
       "      <td>0.019531</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:276138:A:G</td>\n",
       "      <td>16</td>\n",
       "      <td>276138</td>\n",
       "      <td>2.750000e-12</td>\n",
       "      <td>0.586679</td>\n",
       "      <td>0.744141</td>\n",
       "      <td>2.748564e-12</td>\n",
       "      <td>2.192808e-17</td>\n",
       "      <td>-0.000019</td>\n",
       "      <td>-0.275971</td>\n",
       "      <td>0.030206</td>\n",
       "      <td>0.026672</td>\n",
       "      <td>0.027344</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:82511:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>82511</td>\n",
       "      <td>1.460000e-08</td>\n",
       "      <td>0.469891</td>\n",
       "      <td>0.571181</td>\n",
       "      <td>1.458276e-08</td>\n",
       "      <td>6.780125e-15</td>\n",
       "      <td>0.038860</td>\n",
       "      <td>-0.234303</td>\n",
       "      <td>0.030518</td>\n",
       "      <td>0.026027</td>\n",
       "      <td>0.035156</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>page</td>\n",
       "      <td>mean_corp_hgb_conc</td>\n",
       "      <td>chr16:244158:T:C</td>\n",
       "      <td>16</td>\n",
       "      <td>244158</td>\n",
       "      <td>6.630000e-10</td>\n",
       "      <td>0.311663</td>\n",
       "      <td>0.359798</td>\n",
       "      <td>6.625713e-10</td>\n",
       "      <td>9.438585e-01</td>\n",
       "      <td>0.142546</td>\n",
       "      <td>0.146045</td>\n",
       "      <td>0.046537</td>\n",
       "      <td>0.025331</td>\n",
       "      <td>0.964844</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>ukb</td>\n",
       "      <td>ever_smoked</td>\n",
       "      <td>chr8:95301836:T:C</td>\n",
       "      <td>8</td>\n",
       "      <td>95301836</td>\n",
       "      <td>4.170000e-09</td>\n",
       "      <td>0.571303</td>\n",
       "      <td>0.444071</td>\n",
       "      <td>4.175101e-09</td>\n",
       "      <td>9.640772e-01</td>\n",
       "      <td>0.085532</td>\n",
       "      <td>0.084424</td>\n",
       "      <td>0.019743</td>\n",
       "      <td>0.018205</td>\n",
       "      <td>0.972656</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>page</td>\n",
       "      <td>platelet_cnt</td>\n",
       "      <td>chr6:33579153:T:C</td>\n",
       "      <td>6</td>\n",
       "      <td>33579153</td>\n",
       "      <td>3.050000e-10</td>\n",
       "      <td>0.718672</td>\n",
       "      <td>0.739843</td>\n",
       "      <td>3.051170e-10</td>\n",
       "      <td>9.725788e-01</td>\n",
       "      <td>0.106665</td>\n",
       "      <td>0.105793</td>\n",
       "      <td>0.025971</td>\n",
       "      <td>0.017715</td>\n",
       "      <td>0.980469</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>page</td>\n",
       "      <td>platelet_cnt</td>\n",
       "      <td>chr19:16092494:C:A</td>\n",
       "      <td>19</td>\n",
       "      <td>16092494</td>\n",
       "      <td>6.000000e-10</td>\n",
       "      <td>0.019876</td>\n",
       "      <td>0.097543</td>\n",
       "      <td>5.995224e-10</td>\n",
       "      <td>9.846546e-01</td>\n",
       "      <td>-0.171431</td>\n",
       "      <td>-0.169185</td>\n",
       "      <td>0.113732</td>\n",
       "      <td>0.028054</td>\n",
       "      <td>0.988281</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>page</td>\n",
       "      <td>triglycerides</td>\n",
       "      <td>chr8:19965681:T:C</td>\n",
       "      <td>8</td>\n",
       "      <td>19965681</td>\n",
       "      <td>2.540000e-09</td>\n",
       "      <td>0.034727</td>\n",
       "      <td>0.080749</td>\n",
       "      <td>2.537860e-09</td>\n",
       "      <td>9.939309e-01</td>\n",
       "      <td>0.162097</td>\n",
       "      <td>0.162754</td>\n",
       "      <td>0.082243</td>\n",
       "      <td>0.028655</td>\n",
       "      <td>0.996094</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>128 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    study               trait                 SNP  CHROM       POS  \\\n",
       "17    ukb                 MCH    chr16:140238:C:T     16    140238   \n",
       "19    ukb                 MCH    chr16:207212:A:C     16    207212   \n",
       "23    ukb                 MCH    chr16:291988:C:T     16    291988   \n",
       "22    ukb                 MCH    chr16:276138:A:G     16    276138   \n",
       "14    ukb                 MCH     chr16:82511:G:A     16     82511   \n",
       "..    ...                 ...                 ...    ...       ...   \n",
       "96   page  mean_corp_hgb_conc    chr16:244158:T:C     16    244158   \n",
       "53    ukb         ever_smoked   chr8:95301836:T:C      8  95301836   \n",
       "98   page        platelet_cnt   chr6:33579153:T:C      6  33579153   \n",
       "102  page        platelet_cnt  chr19:16092494:C:A     19  16092494   \n",
       "120  page       triglycerides   chr8:19965681:T:C      8  19965681   \n",
       "\n",
       "          PLINK_P    EUR_af    AFR_af    assoc_pval      HET_pval  EUR_beta  \\\n",
       "17   6.540000e-10  0.428832  0.772786  6.542885e-10  1.854274e-24  0.074391   \n",
       "19   2.840000e-09  0.490268  0.844401  2.843981e-09  4.842501e-24  0.049764   \n",
       "23   5.680000e-14  0.177670  0.247776  5.677463e-14  1.097601e-21  0.124169   \n",
       "22   2.750000e-12  0.586679  0.744141  2.748564e-12  2.192808e-17 -0.000019   \n",
       "14   1.460000e-08  0.469891  0.571181  1.458276e-08  6.780125e-15  0.038860   \n",
       "..            ...       ...       ...           ...           ...       ...   \n",
       "96   6.630000e-10  0.311663  0.359798  6.625713e-10  9.438585e-01  0.142546   \n",
       "53   4.170000e-09  0.571303  0.444071  4.175101e-09  9.640772e-01  0.085532   \n",
       "98   3.050000e-10  0.718672  0.739843  3.051170e-10  9.725788e-01  0.106665   \n",
       "102  6.000000e-10  0.019876  0.097543  5.995224e-10  9.846546e-01 -0.171431   \n",
       "120  2.540000e-09  0.034727  0.080749  2.537860e-09  9.939309e-01  0.162097   \n",
       "\n",
       "     AFR_beta  EUR_beta_stderr  AFR_beta_stderr  expected_HET_pval  \\\n",
       "17  -0.277259         0.031295         0.026369           0.003906   \n",
       "19  -0.289335         0.031103         0.028097           0.011719   \n",
       "23  -0.371366         0.042412         0.031606           0.019531   \n",
       "22  -0.275971         0.030206         0.026672           0.027344   \n",
       "14  -0.234303         0.030518         0.026027           0.035156   \n",
       "..        ...              ...              ...                ...   \n",
       "96   0.146045         0.046537         0.025331           0.964844   \n",
       "53   0.084424         0.019743         0.018205           0.972656   \n",
       "98   0.105793         0.025971         0.017715           0.980469   \n",
       "102 -0.169185         0.113732         0.028054           0.988281   \n",
       "120  0.162754         0.082243         0.028655           0.996094   \n",
       "\n",
       "     bonferroni_significant  \n",
       "17                     True  \n",
       "19                     True  \n",
       "23                     True  \n",
       "22                     True  \n",
       "14                     True  \n",
       "..                      ...  \n",
       "96                    False  \n",
       "53                    False  \n",
       "98                    False  \n",
       "102                   False  \n",
       "120                   False  \n",
       "\n",
       "[128 rows x 16 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_assoc.sort_values(\"HET_pval\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fca1ef4f-9086-48b7-b48a-78b0c4ae75ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>trait</th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>PLINK_P</th>\n",
       "      <th>EUR_af</th>\n",
       "      <th>AFR_af</th>\n",
       "      <th>assoc_pval</th>\n",
       "      <th>HET_pval</th>\n",
       "      <th>EUR_beta</th>\n",
       "      <th>AFR_beta</th>\n",
       "      <th>EUR_beta_stderr</th>\n",
       "      <th>AFR_beta_stderr</th>\n",
       "      <th>expected_HET_pval</th>\n",
       "      <th>bonferroni_significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr6:135105435:A:G</td>\n",
       "      <td>6</td>\n",
       "      <td>135105435</td>\n",
       "      <td>2.050000e-08</td>\n",
       "      <td>0.259395</td>\n",
       "      <td>0.068354</td>\n",
       "      <td>2.045087e-08</td>\n",
       "      <td>6.318983e-01</td>\n",
       "      <td>0.187724</td>\n",
       "      <td>0.156203</td>\n",
       "      <td>0.037227</td>\n",
       "      <td>0.056189</td>\n",
       "      <td>0.753906</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:42391:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>42391</td>\n",
       "      <td>2.690000e-09</td>\n",
       "      <td>0.079075</td>\n",
       "      <td>0.476997</td>\n",
       "      <td>2.690165e-09</td>\n",
       "      <td>4.271129e-02</td>\n",
       "      <td>-0.033161</td>\n",
       "      <td>-0.164937</td>\n",
       "      <td>0.061732</td>\n",
       "      <td>0.026178</td>\n",
       "      <td>0.246094</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:46496:A:G</td>\n",
       "      <td>16</td>\n",
       "      <td>46496</td>\n",
       "      <td>2.150000e-10</td>\n",
       "      <td>0.166058</td>\n",
       "      <td>0.217231</td>\n",
       "      <td>2.153261e-10</td>\n",
       "      <td>2.889832e-13</td>\n",
       "      <td>0.070183</td>\n",
       "      <td>-0.319825</td>\n",
       "      <td>0.043447</td>\n",
       "      <td>0.033769</td>\n",
       "      <td>0.058594</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:81470:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>81470</td>\n",
       "      <td>7.100000e-13</td>\n",
       "      <td>0.005474</td>\n",
       "      <td>0.537109</td>\n",
       "      <td>7.100898e-13</td>\n",
       "      <td>3.080980e-01</td>\n",
       "      <td>0.045975</td>\n",
       "      <td>-0.188178</td>\n",
       "      <td>0.229313</td>\n",
       "      <td>0.025934</td>\n",
       "      <td>0.582031</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:82511:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>82511</td>\n",
       "      <td>1.460000e-08</td>\n",
       "      <td>0.469891</td>\n",
       "      <td>0.571181</td>\n",
       "      <td>1.458276e-08</td>\n",
       "      <td>6.780125e-15</td>\n",
       "      <td>0.038860</td>\n",
       "      <td>-0.234303</td>\n",
       "      <td>0.030518</td>\n",
       "      <td>0.026027</td>\n",
       "      <td>0.035156</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:86889:T:C</td>\n",
       "      <td>16</td>\n",
       "      <td>86889</td>\n",
       "      <td>2.800000e-14</td>\n",
       "      <td>0.810523</td>\n",
       "      <td>0.530165</td>\n",
       "      <td>2.804795e-14</td>\n",
       "      <td>2.568922e-04</td>\n",
       "      <td>0.248774</td>\n",
       "      <td>0.126545</td>\n",
       "      <td>0.030218</td>\n",
       "      <td>0.027357</td>\n",
       "      <td>0.128906</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:111939:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>111939</td>\n",
       "      <td>1.280000e-11</td>\n",
       "      <td>0.867701</td>\n",
       "      <td>0.659071</td>\n",
       "      <td>1.282368e-11</td>\n",
       "      <td>3.650972e-10</td>\n",
       "      <td>0.293704</td>\n",
       "      <td>0.097491</td>\n",
       "      <td>0.031784</td>\n",
       "      <td>0.027832</td>\n",
       "      <td>0.082031</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:140238:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>140238</td>\n",
       "      <td>6.540000e-10</td>\n",
       "      <td>0.428832</td>\n",
       "      <td>0.772786</td>\n",
       "      <td>6.542885e-10</td>\n",
       "      <td>1.854274e-24</td>\n",
       "      <td>0.074391</td>\n",
       "      <td>-0.277259</td>\n",
       "      <td>0.031295</td>\n",
       "      <td>0.026369</td>\n",
       "      <td>0.003906</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:161106:T:G</td>\n",
       "      <td>16</td>\n",
       "      <td>161106</td>\n",
       "      <td>1.060000e-11</td>\n",
       "      <td>0.672749</td>\n",
       "      <td>0.814236</td>\n",
       "      <td>1.064191e-11</td>\n",
       "      <td>1.855403e-04</td>\n",
       "      <td>0.242284</td>\n",
       "      <td>0.122894</td>\n",
       "      <td>0.031365</td>\n",
       "      <td>0.028744</td>\n",
       "      <td>0.121094</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:207212:A:C</td>\n",
       "      <td>16</td>\n",
       "      <td>207212</td>\n",
       "      <td>2.840000e-09</td>\n",
       "      <td>0.490268</td>\n",
       "      <td>0.844401</td>\n",
       "      <td>2.843981e-09</td>\n",
       "      <td>4.842501e-24</td>\n",
       "      <td>0.049764</td>\n",
       "      <td>-0.289335</td>\n",
       "      <td>0.031103</td>\n",
       "      <td>0.028097</td>\n",
       "      <td>0.011719</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:244158:T:C</td>\n",
       "      <td>16</td>\n",
       "      <td>244158</td>\n",
       "      <td>1.580000e-09</td>\n",
       "      <td>0.282543</td>\n",
       "      <td>0.345269</td>\n",
       "      <td>1.582692e-09</td>\n",
       "      <td>6.131856e-01</td>\n",
       "      <td>0.128582</td>\n",
       "      <td>0.150102</td>\n",
       "      <td>0.035038</td>\n",
       "      <td>0.028668</td>\n",
       "      <td>0.730469</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:261854:A:G</td>\n",
       "      <td>16</td>\n",
       "      <td>261854</td>\n",
       "      <td>6.300000e-23</td>\n",
       "      <td>0.197080</td>\n",
       "      <td>0.641927</td>\n",
       "      <td>6.298589e-23</td>\n",
       "      <td>1.605620e-12</td>\n",
       "      <td>0.015962</td>\n",
       "      <td>-0.301758</td>\n",
       "      <td>0.040885</td>\n",
       "      <td>0.025023</td>\n",
       "      <td>0.066406</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:276138:A:G</td>\n",
       "      <td>16</td>\n",
       "      <td>276138</td>\n",
       "      <td>2.750000e-12</td>\n",
       "      <td>0.586679</td>\n",
       "      <td>0.744141</td>\n",
       "      <td>2.748564e-12</td>\n",
       "      <td>2.192808e-17</td>\n",
       "      <td>-0.000019</td>\n",
       "      <td>-0.275971</td>\n",
       "      <td>0.030206</td>\n",
       "      <td>0.026672</td>\n",
       "      <td>0.027344</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:291988:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>291988</td>\n",
       "      <td>5.680000e-14</td>\n",
       "      <td>0.177670</td>\n",
       "      <td>0.247776</td>\n",
       "      <td>5.677463e-14</td>\n",
       "      <td>1.097601e-21</td>\n",
       "      <td>0.124169</td>\n",
       "      <td>-0.371366</td>\n",
       "      <td>0.042412</td>\n",
       "      <td>0.031606</td>\n",
       "      <td>0.019531</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:362827:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>362827</td>\n",
       "      <td>4.240000e-18</td>\n",
       "      <td>0.251597</td>\n",
       "      <td>0.631156</td>\n",
       "      <td>4.240675e-18</td>\n",
       "      <td>6.403675e-14</td>\n",
       "      <td>0.024056</td>\n",
       "      <td>-0.287163</td>\n",
       "      <td>0.037271</td>\n",
       "      <td>0.025512</td>\n",
       "      <td>0.050781</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:363523:T:G</td>\n",
       "      <td>16</td>\n",
       "      <td>363523</td>\n",
       "      <td>1.520000e-12</td>\n",
       "      <td>0.383632</td>\n",
       "      <td>0.780864</td>\n",
       "      <td>1.521310e-12</td>\n",
       "      <td>9.770221e-15</td>\n",
       "      <td>0.012013</td>\n",
       "      <td>-0.264971</td>\n",
       "      <td>0.032535</td>\n",
       "      <td>0.026457</td>\n",
       "      <td>0.042969</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:396555:C:G</td>\n",
       "      <td>16</td>\n",
       "      <td>396555</td>\n",
       "      <td>7.570000e-13</td>\n",
       "      <td>0.024901</td>\n",
       "      <td>0.156854</td>\n",
       "      <td>7.575839e-13</td>\n",
       "      <td>4.630959e-02</td>\n",
       "      <td>-0.057018</td>\n",
       "      <td>-0.285327</td>\n",
       "      <td>0.108505</td>\n",
       "      <td>0.038276</td>\n",
       "      <td>0.261719</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:488418:T:A</td>\n",
       "      <td>16</td>\n",
       "      <td>488418</td>\n",
       "      <td>4.580000e-13</td>\n",
       "      <td>0.008807</td>\n",
       "      <td>0.260048</td>\n",
       "      <td>4.574851e-13</td>\n",
       "      <td>3.107068e-01</td>\n",
       "      <td>-0.043783</td>\n",
       "      <td>-0.229266</td>\n",
       "      <td>0.181193</td>\n",
       "      <td>0.031267</td>\n",
       "      <td>0.597656</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:551870:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>551870</td>\n",
       "      <td>7.740000e-09</td>\n",
       "      <td>0.062973</td>\n",
       "      <td>0.480514</td>\n",
       "      <td>7.740229e-09</td>\n",
       "      <td>2.500668e-07</td>\n",
       "      <td>0.190263</td>\n",
       "      <td>-0.182048</td>\n",
       "      <td>0.068972</td>\n",
       "      <td>0.025733</td>\n",
       "      <td>0.097656</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:663204:A:G</td>\n",
       "      <td>16</td>\n",
       "      <td>663204</td>\n",
       "      <td>1.970000e-08</td>\n",
       "      <td>0.357748</td>\n",
       "      <td>0.804225</td>\n",
       "      <td>1.965165e-08</td>\n",
       "      <td>1.537278e-10</td>\n",
       "      <td>0.021040</td>\n",
       "      <td>-0.214873</td>\n",
       "      <td>0.033438</td>\n",
       "      <td>0.026698</td>\n",
       "      <td>0.074219</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:917241:A:G</td>\n",
       "      <td>16</td>\n",
       "      <td>917241</td>\n",
       "      <td>2.070000e-09</td>\n",
       "      <td>0.173019</td>\n",
       "      <td>0.700871</td>\n",
       "      <td>2.072787e-09</td>\n",
       "      <td>2.924312e-03</td>\n",
       "      <td>-0.030660</td>\n",
       "      <td>-0.168368</td>\n",
       "      <td>0.043181</td>\n",
       "      <td>0.025140</td>\n",
       "      <td>0.167969</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:1031151:C:T</td>\n",
       "      <td>16</td>\n",
       "      <td>1031151</td>\n",
       "      <td>6.720000e-10</td>\n",
       "      <td>0.010561</td>\n",
       "      <td>0.317547</td>\n",
       "      <td>6.714206e-10</td>\n",
       "      <td>6.731205e-01</td>\n",
       "      <td>-0.109988</td>\n",
       "      <td>-0.180443</td>\n",
       "      <td>0.165324</td>\n",
       "      <td>0.029181</td>\n",
       "      <td>0.792969</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:2102650:T:C</td>\n",
       "      <td>16</td>\n",
       "      <td>2102650</td>\n",
       "      <td>8.490000e-13</td>\n",
       "      <td>0.169023</td>\n",
       "      <td>0.576460</td>\n",
       "      <td>8.491597e-13</td>\n",
       "      <td>7.938548e-03</td>\n",
       "      <td>-0.067744</td>\n",
       "      <td>-0.194623</td>\n",
       "      <td>0.043976</td>\n",
       "      <td>0.025474</td>\n",
       "      <td>0.199219</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:3912120:G:A</td>\n",
       "      <td>16</td>\n",
       "      <td>3912120</td>\n",
       "      <td>4.420000e-08</td>\n",
       "      <td>0.975267</td>\n",
       "      <td>0.503354</td>\n",
       "      <td>4.424057e-08</td>\n",
       "      <td>1.854332e-08</td>\n",
       "      <td>0.254373</td>\n",
       "      <td>0.071372</td>\n",
       "      <td>0.032507</td>\n",
       "      <td>0.027984</td>\n",
       "      <td>0.089844</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ukb</td>\n",
       "      <td>MCH</td>\n",
       "      <td>chr16:5112819:C:A</td>\n",
       "      <td>16</td>\n",
       "      <td>5112819</td>\n",
       "      <td>5.780000e-09</td>\n",
       "      <td>0.467352</td>\n",
       "      <td>0.018527</td>\n",
       "      <td>5.775389e-09</td>\n",
       "      <td>8.669566e-01</td>\n",
       "      <td>0.171948</td>\n",
       "      <td>0.190320</td>\n",
       "      <td>0.030489</td>\n",
       "      <td>0.106833</td>\n",
       "      <td>0.902344</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   study trait                 SNP  CHROM        POS       PLINK_P    EUR_af  \\\n",
       "10   ukb   MCH  chr6:135105435:A:G      6  135105435  2.050000e-08  0.259395   \n",
       "11   ukb   MCH     chr16:42391:G:A     16      42391  2.690000e-09  0.079075   \n",
       "12   ukb   MCH     chr16:46496:A:G     16      46496  2.150000e-10  0.166058   \n",
       "13   ukb   MCH     chr16:81470:G:A     16      81470  7.100000e-13  0.005474   \n",
       "14   ukb   MCH     chr16:82511:G:A     16      82511  1.460000e-08  0.469891   \n",
       "15   ukb   MCH     chr16:86889:T:C     16      86889  2.800000e-14  0.810523   \n",
       "16   ukb   MCH    chr16:111939:G:A     16     111939  1.280000e-11  0.867701   \n",
       "17   ukb   MCH    chr16:140238:C:T     16     140238  6.540000e-10  0.428832   \n",
       "18   ukb   MCH    chr16:161106:T:G     16     161106  1.060000e-11  0.672749   \n",
       "19   ukb   MCH    chr16:207212:A:C     16     207212  2.840000e-09  0.490268   \n",
       "20   ukb   MCH    chr16:244158:T:C     16     244158  1.580000e-09  0.282543   \n",
       "21   ukb   MCH    chr16:261854:A:G     16     261854  6.300000e-23  0.197080   \n",
       "22   ukb   MCH    chr16:276138:A:G     16     276138  2.750000e-12  0.586679   \n",
       "23   ukb   MCH    chr16:291988:C:T     16     291988  5.680000e-14  0.177670   \n",
       "24   ukb   MCH    chr16:362827:C:T     16     362827  4.240000e-18  0.251597   \n",
       "25   ukb   MCH    chr16:363523:T:G     16     363523  1.520000e-12  0.383632   \n",
       "26   ukb   MCH    chr16:396555:C:G     16     396555  7.570000e-13  0.024901   \n",
       "27   ukb   MCH    chr16:488418:T:A     16     488418  4.580000e-13  0.008807   \n",
       "28   ukb   MCH    chr16:551870:G:A     16     551870  7.740000e-09  0.062973   \n",
       "29   ukb   MCH    chr16:663204:A:G     16     663204  1.970000e-08  0.357748   \n",
       "30   ukb   MCH    chr16:917241:A:G     16     917241  2.070000e-09  0.173019   \n",
       "31   ukb   MCH   chr16:1031151:C:T     16    1031151  6.720000e-10  0.010561   \n",
       "32   ukb   MCH   chr16:2102650:T:C     16    2102650  8.490000e-13  0.169023   \n",
       "33   ukb   MCH   chr16:3912120:G:A     16    3912120  4.420000e-08  0.975267   \n",
       "34   ukb   MCH   chr16:5112819:C:A     16    5112819  5.780000e-09  0.467352   \n",
       "\n",
       "      AFR_af    assoc_pval      HET_pval  EUR_beta  AFR_beta  EUR_beta_stderr  \\\n",
       "10  0.068354  2.045087e-08  6.318983e-01  0.187724  0.156203         0.037227   \n",
       "11  0.476997  2.690165e-09  4.271129e-02 -0.033161 -0.164937         0.061732   \n",
       "12  0.217231  2.153261e-10  2.889832e-13  0.070183 -0.319825         0.043447   \n",
       "13  0.537109  7.100898e-13  3.080980e-01  0.045975 -0.188178         0.229313   \n",
       "14  0.571181  1.458276e-08  6.780125e-15  0.038860 -0.234303         0.030518   \n",
       "15  0.530165  2.804795e-14  2.568922e-04  0.248774  0.126545         0.030218   \n",
       "16  0.659071  1.282368e-11  3.650972e-10  0.293704  0.097491         0.031784   \n",
       "17  0.772786  6.542885e-10  1.854274e-24  0.074391 -0.277259         0.031295   \n",
       "18  0.814236  1.064191e-11  1.855403e-04  0.242284  0.122894         0.031365   \n",
       "19  0.844401  2.843981e-09  4.842501e-24  0.049764 -0.289335         0.031103   \n",
       "20  0.345269  1.582692e-09  6.131856e-01  0.128582  0.150102         0.035038   \n",
       "21  0.641927  6.298589e-23  1.605620e-12  0.015962 -0.301758         0.040885   \n",
       "22  0.744141  2.748564e-12  2.192808e-17 -0.000019 -0.275971         0.030206   \n",
       "23  0.247776  5.677463e-14  1.097601e-21  0.124169 -0.371366         0.042412   \n",
       "24  0.631156  4.240675e-18  6.403675e-14  0.024056 -0.287163         0.037271   \n",
       "25  0.780864  1.521310e-12  9.770221e-15  0.012013 -0.264971         0.032535   \n",
       "26  0.156854  7.575839e-13  4.630959e-02 -0.057018 -0.285327         0.108505   \n",
       "27  0.260048  4.574851e-13  3.107068e-01 -0.043783 -0.229266         0.181193   \n",
       "28  0.480514  7.740229e-09  2.500668e-07  0.190263 -0.182048         0.068972   \n",
       "29  0.804225  1.965165e-08  1.537278e-10  0.021040 -0.214873         0.033438   \n",
       "30  0.700871  2.072787e-09  2.924312e-03 -0.030660 -0.168368         0.043181   \n",
       "31  0.317547  6.714206e-10  6.731205e-01 -0.109988 -0.180443         0.165324   \n",
       "32  0.576460  8.491597e-13  7.938548e-03 -0.067744 -0.194623         0.043976   \n",
       "33  0.503354  4.424057e-08  1.854332e-08  0.254373  0.071372         0.032507   \n",
       "34  0.018527  5.775389e-09  8.669566e-01  0.171948  0.190320         0.030489   \n",
       "\n",
       "    AFR_beta_stderr  expected_HET_pval  bonferroni_significant  \n",
       "10         0.056189           0.753906                   False  \n",
       "11         0.026178           0.246094                   False  \n",
       "12         0.033769           0.058594                    True  \n",
       "13         0.025934           0.582031                   False  \n",
       "14         0.026027           0.035156                    True  \n",
       "15         0.027357           0.128906                    True  \n",
       "16         0.027832           0.082031                    True  \n",
       "17         0.026369           0.003906                    True  \n",
       "18         0.028744           0.121094                    True  \n",
       "19         0.028097           0.011719                    True  \n",
       "20         0.028668           0.730469                   False  \n",
       "21         0.025023           0.066406                    True  \n",
       "22         0.026672           0.027344                    True  \n",
       "23         0.031606           0.019531                    True  \n",
       "24         0.025512           0.050781                    True  \n",
       "25         0.026457           0.042969                    True  \n",
       "26         0.038276           0.261719                   False  \n",
       "27         0.031267           0.597656                   False  \n",
       "28         0.025733           0.097656                    True  \n",
       "29         0.026698           0.074219                    True  \n",
       "30         0.025140           0.167969                   False  \n",
       "31         0.029181           0.792969                   False  \n",
       "32         0.025474           0.199219                   False  \n",
       "33         0.027984           0.089844                    True  \n",
       "34         0.106833           0.902344                   False  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_assoc[df_assoc.trait.isin([\"MCH\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3bfca74e-963a-46db-88bb-529f81b2d69e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# all SNPs\n",
      "128 SNPs\n",
      "Pearson's r = 0.73, SE=0.04\n",
      "Deming regression slope: 1.36. Bootstrap mean (se): 1.37 (0.13)\n",
      "lambda gc = 3.02, ci=[2.27, 3.53]\n",
      "\n",
      "# SNPs removing MCH\n",
      "103 SNPs\n",
      "Pearson's r = 0.84, SE=0.03\n",
      "Deming regression slope: 1.12. Bootstrap mean (se): 1.12 (0.05)\n",
      "lambda gc = 2.64, ci=[1.59, 3.21]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"# all SNPs\")\n",
    "report_assoc(df_assoc)\n",
    "print(\"\")\n",
    "print(\"# SNPs removing MCH\")\n",
    "report_assoc(df_assoc[~df_assoc.trait.isin([\"MCH\"])])\n",
    "print(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ec9f4d2-f48b-4141-b23e-f99dd7a667d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4, 2), dpi=150)\n",
    "sns.scatterplot(\n",
    "    x=df_assoc.EUR_af - df_assoc.AFR_af,\n",
    "    y=-np.log10(df_assoc.HET_pval),\n",
    "    hue=df_assoc.hue,\n",
    "    s=4,\n",
    "    linewidth=0,\n",
    "    palette=\"tab10\",\n",
    "    ax=ax,\n",
    ")\n",
    "\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "ax.legend(handles=handles, labels=labels, loc=\"best\", fontsize=6)\n",
    "\n",
    "ax.set_xlabel(\"EUR freq - AFR freq\")\n",
    "ax.set_ylabel(\"HET $-\\log_{10}(p)$\")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(4, 2), dpi=150)\n",
    "sns.scatterplot(\n",
    "    x=df_assoc.EUR_beta - df_assoc.AFR_beta,\n",
    "    y=-np.log10(df_assoc.HET_pval),\n",
    "    hue=df_assoc.hue,\n",
    "    s=4,\n",
    "    linewidth=0,\n",
    "    palette=\"tab10\",\n",
    "    ax=ax,\n",
    ")\n",
    "\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "ax.legend(handles=handles, labels=labels, loc=\"best\", fontsize=6)\n",
    "\n",
    "ax.set_xlabel(\"EUR beta - AFR beta\")\n",
    "ax.set_ylabel(\"HET $-\\log_{10}(p)$\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95c9b621-caf6-4588-957d-fe9c1676f44f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def qqplot(pval, ax=None):\n",
    "    if ax is None:\n",
    "        ax = plt.gca()\n",
    "    expected_pval = stats.norm.sf(quantile_normalize(-pval))\n",
    "    ax.scatter(-np.log10(expected_pval), -np.log10(pval), s=2)\n",
    "    lim = max(-np.log10(expected_pval))\n",
    "    ax.plot([0, lim], [0, lim], \"r--\")\n",
    "    ax.set_xlabel(\"Expected -$\\log_{10}(p)$\")\n",
    "    ax.set_ylabel(\"Observed -$\\log_{10}(p)$\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffc42925-154a-4085-a0dc-1ee20657e1b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# color points by trait after Bonferroni correction\n",
    "df_plot = df_assoc[[\"trait_id\", \"SNPS\", \"HET_pval\", \"assoc_pval\"]].copy()\n",
    "df_plot.loc[:, \"expected_HET_pval\"] = stats.norm.sf(\n",
    "    quantile_normalize(-df_plot.HET_pval)\n",
    ")\n",
    "df_plot.loc[:, \"bonferroni_significant\"] = df_plot[\"HET_pval\"] < 0.05 / len(df_plot)\n",
    "\n",
    "df_plot[\"hue\"] = \"\"\n",
    "df_plot.loc[df_plot.bonferroni_significant, \"hue\"] = df_plot.loc[\n",
    "    df_plot.bonferroni_significant, \"trait_id\"\n",
    "]\n",
    "df_plot.loc[~df_plot.bonferroni_significant, \"hue\"] = \"non-sig\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32fde553-919b-4ad0-b6a1-685ef685bcb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4, 4), dpi=150)\n",
    "sns.scatterplot(\n",
    "    x=-np.log10(df_plot.expected_HET_pval),\n",
    "    y=-np.log10(df_plot.HET_pval),\n",
    "    hue=df_plot.hue,\n",
    "    s=10,\n",
    "    linewidth=0,\n",
    "    palette=\"tab10\",\n",
    "    ax=ax,\n",
    ")\n",
    "lim = max(-np.log10(df_plot.expected_HET_pval))\n",
    "ax.plot([0, lim], [0, lim], \"r--\")\n",
    "ax.set_xlabel(\"Expected -$\\log_{10}(p)$\")\n",
    "ax.set_ylabel(\"Observed -$\\log_{10}(p)$\")\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "ax.set_title(\"Heterogeneity p-value at index SNPs\")\n",
    "\n",
    "ax.legend(handles=handles, labels=labels, loc=\"best\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
