{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4c71feda-724a-49b0-96e9-ad0a7075d76f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "The lab_black extension is already loaded. To reload it, use:\n",
      "  %reload_ext lab_black\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "\n",
    "import xarray as xr\n",
    "import numpy as np\n",
    "import dapgen\n",
    "import pandas as pd\n",
    "import dask.array as da\n",
    "import itertools\n",
    "import submitit\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import admix_genet_cor\n",
    "import pandas as pd\n",
    "import admix\n",
    "from os.path import join\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afc8e62e-5355-40f4-b76f-017d25b8cba7",
   "metadata": {},
   "source": [
    "# Overview of simulation studies\n",
    "- p_causal: 0.001, 1.0\n",
    "- var_g: 1.0\n",
    "- var_e: 1.0\n",
    "- gamma: 0.5, 0.8, 1.0\n",
    "- set of SNPs: simulate from the imputed SNPs, the non-zero effects are simulated from SNPs with allele frequencies > 0.005 in both populations.\n",
    "\n",
    "From S-LDXR paper: \n",
    "```\n",
    "In these simulations, we randomly selected 10% of the SNPs to be causal in each population, with 80% of causal variants in each population shared with the other population, and sampled perfectly correlated causal effect sizes for shared causal variants using Eq. In these simulations, we set the variance of causal effect size of each SNP j in both populations to be proportional to $[p_{j,max}(1−p_{j,max})]^\\alpha$, where pj,max is the maximum MAF of SNP j in the two populations.  We set α to − 0.38, as previously estimated for 25 UK Biobank diseases \n",
    "and complex traits in ref.\n",
    "```\n",
    "\n",
    "Here, we simulate from the imputed SNPs the non-zero effects are simulated from SNPs with allele frequencies > 0.005 in both populations. And from these, we sample 1% of causal effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b78dd1f9-7e8d-4c0f-ad57-553122ea893b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONSTANTS\n",
    "\n",
    "CHROM = 22\n",
    "DATA_ROOT_DIR = (\n",
    "    \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01_dataset/out/aframr\"\n",
    ")\n",
    "\n",
    "# define the simulation parameters\n",
    "\n",
    "df_simulate_params = pd.DataFrame(\n",
    "    [\n",
    "        params\n",
    "        for params in itertools.product(\n",
    "            [0.05, 0.1],\n",
    "            [0.001, 0.01],\n",
    "            [0.8, 1.0],\n",
    "            [\"mafukb\"],\n",
    "        )\n",
    "    ],\n",
    "    columns=[\"hsq\", \"pcausal\", \"cor\", \"hermodel\"],\n",
    ")\n",
    "df_simulate_params[\"out_prefix\"] = df_simulate_params.apply(\n",
    "    lambda row: f\"out/pheno/hsq-{row.hsq}-pcausal-{row.pcausal}-cor-{row.cor}-hermodel-{row.hermodel}\",\n",
    "    axis=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b23a3f5-4ebc-4b32-9cbf-e6b9dec8a060",
   "metadata": {},
   "source": [
    "# Step 1: Simulate phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "89ae1a5c-d064-43a1-bab7-acc6873e26e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_snp_prior_var(df_snp_info, her_model):\n",
    "    \"\"\"\n",
    "    Calculate the SNP prior variance from SNP information\n",
    "    \"\"\"\n",
    "    assert her_model in [\"uniform\", \"gcta\", \"ldak\", \"mafukb\"]\n",
    "    if her_model == \"uniform\":\n",
    "        return np.ones(len(df_snp_info))\n",
    "    elif her_model == \"gcta\":\n",
    "        freq = df_snp_info[\"FREQ\"].values\n",
    "        assert np.all(freq > 0), \"frequencies should be larger than zero\"\n",
    "        return np.float_power(freq * (1 - freq), -1)\n",
    "    elif her_model == \"mafukb\":\n",
    "        # MAF-dependent genetic architecture, \\alpha = -0.38 estimated from meta-analysis in UKB traits\n",
    "        freq = df_snp_info[\"FREQ\"].values\n",
    "        assert np.all(freq > 0), \"frequencies should be larger than zero\"\n",
    "        return np.float_power(freq * (1 - freq), -0.38)\n",
    "    elif her_model == \"ldak\":\n",
    "        freq, weight = df_snp_info[\"FREQ\"].values, df_snp_info[\"LDAK_WEIGHT\"].values\n",
    "        return np.float_power(freq * (1 - freq), -0.25) * weight\n",
    "    else:\n",
    "        raise NotImplementedError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "54e09928-ccda-4f45-9749-c0d06d72cf81",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_simulate_pheno(\n",
    "    hsq: float,\n",
    "    pcausal: float,\n",
    "    cor: float,\n",
    "    hermodel: str,\n",
    "    out_prefix: str,\n",
    "):\n",
    "    \"\"\"\n",
    "    her_model: one of [uniform, gcta, ldak]\n",
    "\n",
    "    \"\"\"\n",
    "    N_SIM = 30\n",
    "    N_ANC = 2\n",
    "\n",
    "    np.random.seed(1234)\n",
    "\n",
    "    pfile = f\"{DATA_ROOT_DIR}/imputed/chr{CHROM}\"\n",
    "    print(\n",
    "        f\"hsq: {hsq}, pcausal: {pcausal}, cor: {cor}, hermodel: {hermodel}, out_prefix: {out_prefix}\"\n",
    "    )\n",
    "    geno, df_snp, df_indiv = dapgen.read_pfile(pfile, phase=True, snp_chunk=1024)\n",
    "    lanc = admix.io.read_lanc(pfile + \".lanc\", snp_chunk=1024)\n",
    "    df_snp_info = pd.read_csv(pfile + \".snp_info\", sep=\"\\t\")\n",
    "    snp_prior_var = calc_snp_prior_var(df_snp_info, hermodel)\n",
    "\n",
    "    assert np.all(df_snp_info.SNP == df_snp.index.values)\n",
    "\n",
    "    # simulate effects\n",
    "    snp_subset = np.where(\n",
    "        (df_snp_info.EUR_FREQ > 0.005) & (df_snp_info.AFR_FREQ > 0.005)\n",
    "    )[0]\n",
    "    beta = np.zeros((df_snp_info.shape[0], N_ANC, N_SIM))  # (n_snp, n_anc, n_sim)\n",
    "    n_causal = int(len(snp_subset) * pcausal)\n",
    "    for i_sim in range(N_SIM):\n",
    "        cau = sorted(np.random.choice(snp_subset, size=n_causal, replace=False))\n",
    "\n",
    "        i_beta = np.random.multivariate_normal(\n",
    "            mean=[0.0, 0.0],\n",
    "            cov=np.array([[1, cor], [cor, 1]]),\n",
    "            size=n_causal,\n",
    "        )\n",
    "\n",
    "        i_beta = i_beta * np.sqrt(snp_prior_var[cau])[:, None]\n",
    "\n",
    "        for i_anc in range(N_ANC):\n",
    "            beta[cau, i_anc, i_sim] = i_beta[:, i_anc]\n",
    "\n",
    "    sim = admix_genet_cor.simulate_continuous_pheno(\n",
    "        geno=geno, lanc=lanc, hsq=hsq, beta=beta, n_sim=N_SIM, apa_center=False\n",
    "    )\n",
    "    np.save(out_prefix + \".beta.npy\", sim[\"beta\"])\n",
    "    df_pheno = pd.DataFrame(\n",
    "        sim[\"pheno\"],\n",
    "        index=df_indiv.index,\n",
    "        columns=[f\"SIM_{i}\" for i in range(sim[\"pheno\"].shape[1])],\n",
    "    )\n",
    "    df_pheno.to_csv(out_prefix + \".pheno.tsv\", index=True, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "27a2381d-6e93-4073-afe3-ad3c07ae3139",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=250,\n",
    "    memory_g=40,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_simulate_pheno,\n",
    "    df_simulate_params.hsq,\n",
    "    df_simulate_params.pcausal,\n",
    "    df_simulate_params.cor,\n",
    "    df_simulate_params.hermodel,\n",
    "    df_simulate_params.out_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f98c3a15-9701-495b-a09b-8fed964d47fe",
   "metadata": {},
   "source": [
    "# Step 2: Compute GRM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3719f631-3cf3-4515-b6fe-b0136436dd20",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_compute_grm(hermodel: str, out_prefix: str):\n",
    "\n",
    "    pfile = f\"{DATA_ROOT_DIR}/imputed/chr{CHROM}\"\n",
    "\n",
    "    geno, df_snp, df_indiv = dapgen.read_pfile(pfile, phase=True, snp_chunk=512)\n",
    "    lanc = admix.io.read_lanc(pfile + \".lanc\", snp_chunk=512)\n",
    "\n",
    "    df_snp_info = pd.read_csv(pfile + \".snp_info\", sep=\"\\t\")\n",
    "    snp_prior_var = calc_snp_prior_var(df_snp_info, hermodel)\n",
    "    print(\"snp_prior_var:\")\n",
    "    print(snp_prior_var)\n",
    "\n",
    "    snp_subset = np.where(\n",
    "        (df_snp_info.EUR_FREQ > 0.005) & (df_snp_info.AFR_FREQ > 0.005)\n",
    "    )[0]\n",
    "\n",
    "    K1, K2, K12 = admix_genet_cor.compute_grm(\n",
    "        geno=geno[snp_subset, :, :],\n",
    "        lanc=lanc[snp_subset, :, :],\n",
    "        snp_prior_var=snp_prior_var[snp_subset],\n",
    "        apa_center=True,\n",
    "    )\n",
    "\n",
    "    np.save(out_prefix + \".K1.npy\", K1)\n",
    "    np.save(out_prefix + \".K2.npy\", K2)\n",
    "    np.save(out_prefix + \".K12.npy\", K12)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6468afed-6298-46d7-a338-8da43e2ce0ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=180,\n",
    "    memory_g=40,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "df_grm_params = pd.DataFrame(\n",
    "    [\n",
    "        params\n",
    "        for params in itertools.product(\n",
    "            [\"mafukb\"],\n",
    "        )\n",
    "    ],\n",
    "    columns=[\"hermodel\"],\n",
    ")\n",
    "\n",
    "df_grm_params[\"out_prefix\"] = df_grm_params.apply(\n",
    "    lambda row: f\"out/grm/hermodel-{row.hermodel}\",\n",
    "    axis=1,\n",
    ")\n",
    "jobs = executor.map_array(\n",
    "    submit_compute_grm,\n",
    "    df_grm_params.hermodel,\n",
    "    df_grm_params.out_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24fadf06-d0bd-497e-9870-5923706bcf7a",
   "metadata": {},
   "source": [
    "# Step 3: Estimation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e49614c8-c2cf-4fdb-b5a5-a228898a7a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_estimate_params = df_simulate_params.copy()\n",
    "df_estimate_params[\"estimate_out_prefix\"] = df_estimate_params.apply(\n",
    "    lambda row: row.out_prefix.replace(\"pheno\", \"estimate\"), axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2684703d-fd74-4d4f-aa98-272d885119bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_estimate(hermodel: str, pheno: str, out_prefix: str):\n",
    "    \"\"\"\n",
    "    snp_set: set of SNPs used as GRM, for estimation\n",
    "    pheno: phenotype file\n",
    "    out_prefix: prefix to the output\n",
    "    \"\"\"\n",
    "    print(f\"submit_estimate:\\npheno={pheno}, out_prefix={out_prefix}\")\n",
    "    # Step 1: load GRM\n",
    "    grm_prefix = f\"out/grm/hermodel-{hermodel}\"\n",
    "    K1 = np.load(grm_prefix + \".K1.npy\")\n",
    "    K2 = np.load(grm_prefix + \".K2.npy\")\n",
    "    K12 = np.load(grm_prefix + \".K12.npy\")\n",
    "\n",
    "    # step 2 load data\n",
    "\n",
    "    A1 = K1 + K2\n",
    "    A2 = K12 + K12.T\n",
    "\n",
    "    # K1, K2, K12 are no longer used, release memory for them\n",
    "    del K1, K2, K12\n",
    "\n",
    "    # Step 2: load phenotype and estimate\n",
    "    pheno = pd.read_csv(pheno, delim_whitespace=True, index_col=0)\n",
    "    print(pheno.head())\n",
    "    n_indiv = pheno.shape[0]\n",
    "\n",
    "    rls_list = admix_genet_cor.estimate_genetic_cor(\n",
    "        A1, A2, pheno.values, cov=np.ones((n_indiv, 1))\n",
    "    )\n",
    "\n",
    "    if len(rls_list[0]) == 2:\n",
    "        data = [\n",
    "            {\"estimate\": rls[0].tolist(), \"varcov\": rls[1].tolist()} for rls in rls_list\n",
    "        ]\n",
    "    else:\n",
    "        data = [{\"estimate\": rls.tolist()} for rls in rls_list]\n",
    "\n",
    "    with open(out_prefix + \".estimate.json\", \"w\") as out:\n",
    "        json.dump(data, out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1204590c-6f2f-4d71-9675-07098239eba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=30,\n",
    "    memory_g=48,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_estimate,\n",
    "    df_estimate_params.hermodel,\n",
    "    (df_estimate_params[\"out_prefix\"] + \".pheno.tsv\"),\n",
    "    df_estimate_params.estimate_out_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ed81e16c-9296-4272-be3a-bd08fd7d47ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SgeJob<job_id=449371_1, task_id=0, state=\"pending\">,\n",
       " SgeJob<job_id=449371_2, task_id=0, state=\"pending\">,\n",
       " SgeJob<job_id=449371_3, task_id=0, state=\"pending\">,\n",
       " SgeJob<job_id=449371_4, task_id=0, state=\"pending\">]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jobs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09a10190-1675-454a-904a-3fa5ea2833a3",
   "metadata": {},
   "source": [
    "# LEGACY versions\n",
    "\n",
    "def submit_simulate_pheno(\n",
    "    hsq: float,\n",
    "    pcausal: float,\n",
    "    cor: float,\n",
    "    her_model: str,\n",
    "    out_prefix: str,\n",
    "):\n",
    "    \"\"\"\n",
    "    her_model: one of [uniform, gcta, ldak]\n",
    "\n",
    "    \"\"\"\n",
    "    assert her_model in [\"uniform\", \"gcta\", \"ldak\"]\n",
    "\n",
    "    np.random.seed(1234)\n",
    "\n",
    "    pfile = f\"{DATA_ROOT_DIR}/imputed/chr{CHROM}\"\n",
    "    lanc_path = pfile + \".lanc\"\n",
    "    print(\n",
    "        f\"hsq: {hsq}, pcausal: {pcausal}, cor: {cor}, hermodel: {her_model}, out_prefix: {out_prefix}\"\n",
    "    )\n",
    "    geno, df_snp, df_indiv = dapgen.read_pfile(pfile, phase=True, snp_chunk=1024)\n",
    "    lanc = admix.io.read_lanc(lanc_path, snp_chunk=1024)\n",
    "\n",
    "    df_snp_info = pd.read_csv(pfile + \".snp_info\", sep=\"\\t\")\n",
    "    snp_prior_var = calc_snp_prior_var(df_snp_info, her_model)\n",
    "\n",
    "    print(snp_prior_var)\n",
    "\n",
    "    print(f\"n_causal={int(geno.shape[0] * pcausal)}\")\n",
    "\n",
    "    sim = admix_genet_cor.simulate_continuous_pheno(\n",
    "        geno=geno,\n",
    "        lanc=lanc,\n",
    "        hsq=hsq,\n",
    "        cor=cor,\n",
    "        n_causal=int(geno.shape[0] * pcausal),\n",
    "        snp_prior_var=snp_prior_var,\n",
    "        n_sim=30,\n",
    "    )\n",
    "    np.save(out_prefix + \".beta.npy\", sim[\"beta\"])\n",
    "    df_pheno = pd.DataFrame(\n",
    "        sim[\"pheno\"],\n",
    "        index=df_indiv.index,\n",
    "        columns=[f\"SIM_{i}\" for i in range(sim[\"pheno\"].shape[1])],\n",
    "    )\n",
    "    df_pheno.to_csv(out_prefix + \".pheno.tsv\", index=True, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1531923-fe81-44f3-b363-82466463f9cb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
