{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5b023002-29e7-484b-a74f-df078db769c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-09-23 18:57:46--  https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-019-1310-4/MediaObjects/41586_2019_1310_MOESM3_ESM.xlsx\n",
      "Resolving static-content.springer.com (static-content.springer.com)... 151.101.196.95\n",
      "Connecting to static-content.springer.com (static-content.springer.com)|151.101.196.95|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 1887821 (1.8M) [application/octet-stream]\n",
      "Saving to: ‘page-supp-tables.xlsx’\n",
      "\n",
      "100%[======================================>] 1,887,821   --.-K/s   in 0.04s   \n",
      "\n",
      "2021-09-23 18:57:46 (40.9 MB/s) - ‘page-supp-tables.xlsx’ saved [1887821/1887821]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-019-1310-4/MediaObjects/41586_2019_1310_MOESM3_ESM.xlsx -O page-supp-tables.xlsx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6b9a3a36-9d68-463b-b58a-cb58a5932b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "\n",
    "import xarray as xr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import admix\n",
    "import dask\n",
    "import dask.array as da\n",
    "from tqdm import tqdm\n",
    "import statsmodels.api as sm\n",
    "from scipy.stats import pearsonr, ttest_ind\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from admix.data import quantile_normalize\n",
    "\n",
    "import requests\n",
    "import json\n",
    "import sys\n",
    "import pickle\n",
    "import vcf\n",
    "\n",
    "sys.path.append(\"../\")\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e58bee3-f8f8-421e-b28c-198385e68b43",
   "metadata": {},
   "source": [
    "# Loci analysis\n",
    "1. Take the top SNPs for each significant loci\n",
    "2. Test the heterogeneity score.\n",
    "3. Aggregate over multiple traits and multiple regions and show QQ-plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7d362c3-3726-4cdc-89e5-278b7efb5570",
   "metadata": {},
   "source": [
    "Also see \n",
    "- https://github.com/gokceneraslan/opentargets-genetics-python\n",
    "- https://community.opentargets.org/t/how-to-access-finngen-gwas-data-using-the-open-targets-genetics-portal-api/254/4\n",
    "- https://api.genetics.opentargets.org/graphql/schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "160fc588-82fd-4ea3-96e8-db07caf0f811",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(URL_SUPP_TABLE, sheet_name=\"ST5 All known SNPs by trait\", skiprows=1)\n",
    "df = df.loc[~pd.isna(df[\"Chromosome\"]), :].astype({\"Chromosome\": int, \"Pos_hg19\": int})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8699d022-ee33-48ab-8732-be48290157d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "study_info_query = \"\"\"\n",
    "    query test ($myStudyId: String! ) {\n",
    "      studyInfo(studyId: $myStudyId) {\n",
    "          traitReported\n",
    "          source\n",
    "          traitEfos\n",
    "          pmid\n",
    "          pubAuthor\n",
    "          ancestryInitial\n",
    "          numAssocLoci\n",
    "          nTotal\n",
    "      }\n",
    "    }\n",
    "\"\"\"\n",
    "\n",
    "manhattan_query = \"\"\"\n",
    "    query test ($myStudyId: String! ) {\n",
    "      manhattan(studyId: $myStudyId) {\n",
    "        associations {\n",
    "          variant {\n",
    "            id\n",
    "            rsId\n",
    "            chromosome\n",
    "            position\n",
    "            nearestCodingGene {\n",
    "              id\n",
    "              symbol\n",
    "            }\n",
    "            nearestCodingGeneDistance\n",
    "          }\n",
    "          pval\n",
    "          credibleSetSize\n",
    "          ldSetSize\n",
    "          oddsRatio\n",
    "          oddsRatioCILower\n",
    "          oddsRatioCIUpper\n",
    "          beta\n",
    "          betaCILower\n",
    "          betaCIUpper\n",
    "          direction\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9cc4e9c6-37d7-4289-8364-72697ac040da",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate(apa, beta, cov):\n",
    "    cov_effects = np.random.normal(loc=0, scale=0.1, size=cov.shape[1])\n",
    "    y = (\n",
    "        np.dot(apa, beta)\n",
    "        + np.dot(cov, cov_effects)\n",
    "        + np.random.normal(size=apa.shape[0])\n",
    "    )\n",
    "    return y\n",
    "\n",
    "\n",
    "def test_het(apa, y, cov):\n",
    "    design = sm.add_constant(np.hstack([apa, cov]))\n",
    "    model = sm.OLS(y, design).fit()\n",
    "\n",
    "    A = np.zeros([1, len(model.params)])\n",
    "    A[0, 1] = 1\n",
    "    A[0, 2] = -1\n",
    "    p_ftest = model.f_test(A).pvalue.item()\n",
    "    return p_ftest, model\n",
    "\n",
    "\n",
    "def test_assoc(apa, y, cov):\n",
    "    design = sm.add_constant(np.hstack([apa.sum(axis=1)[:, np.newaxis], cov]))\n",
    "    model = sm.OLS(y, design).fit()\n",
    "    return model.pvalues[1], model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bf991eda-8df0-42f6-ac81-bc2d67e25735",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPP_TABLE_URL = \"https://www.dropbox.com/s/jck2mhjby2ur55j/supp_tables.xlsx?dl=1\"\n",
    "trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"trait-info\")\n",
    "trait_list = trait_info[\"trait\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "16f0ca53-1015-4f27-8c98-d93e29a08d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_url = \"https://api.genetics.opentargets.org/graphql\"\n",
    "\n",
    "df_assoc_list = []\n",
    "for _, row in trait_info.iterrows():\n",
    "    variables = {\"myStudyId\": row[\"open-targets-studyid\"]}\n",
    "    r = requests.post(base_url, json={\"query\": manhattan_query, \"variables\": variables})\n",
    "\n",
    "    response = json.loads(r.text)\n",
    "    assoc = response[\"data\"][\"manhattan\"][\"associations\"]\n",
    "\n",
    "    df_assoc = dict()\n",
    "    # variant information\n",
    "    for v in [\n",
    "        \"id\",\n",
    "        \"rsId\",\n",
    "        \"chromosome\",\n",
    "        \"nearestCodingGene\",\n",
    "        \"nearestCodingGeneDistance\",\n",
    "    ]:\n",
    "        if v == \"nearestCodingGene\":\n",
    "            df_assoc[v] = [d[\"variant\"][v][\"symbol\"] for d in assoc]\n",
    "        else:\n",
    "            df_assoc[v] = [d[\"variant\"][v] for d in assoc]\n",
    "    for v in [\n",
    "        \"pval\",\n",
    "        \"credibleSetSize\",\n",
    "        \"ldSetSize\",\n",
    "        \"oddsRatio\",\n",
    "        \"oddsRatioCILower\",\n",
    "        \"oddsRatioCIUpper\",\n",
    "        \"beta\",\n",
    "        \"betaCILower\",\n",
    "        \"betaCIUpper\",\n",
    "        \"direction\",\n",
    "    ]:\n",
    "        df_assoc[v] = [d[v] for d in assoc]\n",
    "    df_assoc = pd.DataFrame(df_assoc)\n",
    "    df_assoc.insert(\n",
    "        3, \"position\", value=df_assoc[\"id\"].apply(lambda x: int(x.split(\"_\")[1]))\n",
    "    )\n",
    "    df_assoc.insert(0, \"trait\", value=row.trait)\n",
    "    df_assoc_list.append(df_assoc)\n",
    "df_assoc = pd.concat(df_assoc_list)\n",
    "df_assoc = df_assoc.drop(\n",
    "    columns=[\n",
    "        \"oddsRatio\",\n",
    "        \"oddsRatioCILower\",\n",
    "        \"oddsRatioCIUpper\",\n",
    "        \"beta\",\n",
    "        \"betaCILower\",\n",
    "        \"betaCIUpper\",\n",
    "        \"direction\",\n",
    "        \"credibleSetSize\",\n",
    "    ]\n",
    ").rename(columns={\"pval\": \"GWAS_catalog_pval\"})\n",
    "df_assoc = df_assoc.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "38d802ea-7f7b-491f-856d-074a8d183042",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 22/22 [00:03<00:00,  7.11it/s]\n"
     ]
    }
   ],
   "source": [
    "dset = load_hm3()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "12a3f18b-0833-424f-b716-45a1431c7f36",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  4%|▍         | 1/24 [00:16<06:12, 16.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_159204893_T_C not available in data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  8%|▊         | 2/24 [00:48<09:27, 25.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11_5227002_T_A not available in data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 62%|██████▎   | 15/24 [03:56<01:43, 11.52s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21_34455875_G_A not available in data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 67%|██████▋   | 16/24 [04:04<01:26, 10.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6_36678991_C_CTA not available in data\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 24/24 [06:22<00:00, 15.95s/it]\n"
     ]
    }
   ],
   "source": [
    "DATA_DIR = \"/u/project/sgss/PAGE/ImputedGWAS_topmedfrz8/\"\n",
    "\n",
    "dict_record = {}\n",
    "for _, row in df_assoc.iterrows():\n",
    "    chrom = row.chromosome\n",
    "    position = row.position\n",
    "    vcf_reader = vcf.Reader(filename=join(DATA_DIR, f\"MEGA_all.chr{chrom}.filtered.vcf.gz\"))\n",
    "    try:\n",
    "        record = next(vcf_reader.fetch(f\"chr{chrom}\", position - 1, position))\n",
    "        dict_record[row.id] = record\n",
    "    except StopIteration:\n",
    "        print(f\"{row.id} not available in data\")\n",
    "\n",
    "dict_hit_info = {}\n",
    "for snp in dict_record:\n",
    "    record = dict_record[snp]\n",
    "    dict_hit_info[snp] = {\"gt\": np.array([sample.gt_alleles for sample in record.samples], dtype=int),\n",
    "                          \"sample\": [sample.sample for sample in record.samples],\n",
    "                          \"info\": record.INFO}\n",
    "    \n",
    "with open(\"out/locus_het/gwas_hit_info.pkl\", 'wb') as f:\n",
    "    pickle.dump(dict_hit_info, f)\n",
    "    \n",
    "# with open(\"data/gwas_hit_info.pkl\", 'rb') as f:\n",
    "#     dict_hit_info = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "abd59d96-4c1b-4c61-874d-d45117c43333",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"out/locus_het/gwas_hit_info.pkl\", \"rb\") as f:\n",
    "    dict_hit_info = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3051bc8f-8aaf-4241-8298-d5bd5970e504",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "179it [00:00, 217.99it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<xarray.backends.zarr.ZarrStore at 0x2af0a3d649e0>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt = np.dstack([dict_hit_info[snp][\"gt\"] for snp in dict_hit_info])\n",
    "sample = [dict_hit_info[snp][\"sample\"] for snp in dict_hit_info]\n",
    "assert all(x == sample[0] for x in sample)\n",
    "sample = sample[0]\n",
    "\n",
    "R2 = [dict_hit_info[snp][\"info\"][\"R2\"] for snp in dict_hit_info]\n",
    "AF = [dict_hit_info[snp][\"info\"][\"AF\"] for snp in dict_hit_info]\n",
    "\n",
    "dset_gwas_hit = xr.Dataset(\n",
    "    data_vars={\n",
    "        \"geno\": (\n",
    "            (\"indiv\", \"snp\", \"ploidy\"),\n",
    "            da.from_array(np.swapaxes(gt, 1, 2), chunks=-1),\n",
    "        ),\n",
    "    },\n",
    "    coords={\n",
    "        \"snp\": np.array(list(dict_hit_info.keys())).astype(str),\n",
    "        \"indiv\": np.array(sample).astype(str),\n",
    "        \"R2@snp\": (\"snp\", np.array(R2, dtype=float)),\n",
    "        \"AF@snp\": (\"snp\", np.array(AF, dtype=float)),\n",
    "    },\n",
    "    attrs={\"n_anc\": 2},\n",
    ")\n",
    "\n",
    "info = [s.split(\"_\") for s in dset_gwas_hit.snp.values]\n",
    "dset_gwas_hit = dset_gwas_hit.assign_coords(\n",
    "    {\n",
    "        \"CHROM@snp\": (\"snp\", [int(i[0]) for i in info]),\n",
    "        \"POS@snp\": (\"snp\", [int(i[1]) for i in info]),\n",
    "        \"REF@snp\": (\"snp\", [str(i[2]) for i in info]),\n",
    "        \"ALT@snp\": (\"snp\", [str(i[3]) for i in info]),\n",
    "    }\n",
    ")\n",
    "\n",
    "# align `dset_gwas_hit to `dset`\n",
    "dset_gwas_hit = dset_gwas_hit.sel(indiv=dset.indiv.values)\n",
    "\n",
    "# fill in lanc\n",
    "df_dset = dset.snp.to_dataframe().reset_index(drop=True)\n",
    "snp_index = []\n",
    "for _, snp in tqdm(dset_gwas_hit.snp.to_dataframe().iterrows()):\n",
    "    chrom, pos = snp[\"CHROM@snp\"], snp[\"POS@snp\"]\n",
    "    df_tmp = df_dset[df_dset[\"CHROM@snp\"] == chrom]\n",
    "    snp_index.append((df_tmp[\"POS@snp\"] - pos).abs().idxmin())\n",
    "lanc = dset.isel(snp=snp_index).lanc.values\n",
    "dset_gwas_hit = dset_gwas_hit.assign(\n",
    "    lanc=((\"indiv\", \"snp\", \"ploidy\"), da.from_array(lanc, chunks=-1))\n",
    ")\n",
    "\n",
    "# fill in individual information\n",
    "for col in dset:\n",
    "    if col.endswith(\"@indiv\"):\n",
    "        dset_gwas_hit.coords[col] = (\"indiv\", dset[col].values)\n",
    "admix.tools.allele_per_anc(dset_gwas_hit)\n",
    "dset_gwas_hit.to_zarr(\"out/locus_het/gwas_hit.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "580a6016-c139-424a-bced-5cb91043ef3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_assoc[\"EUR_af\"] = np.nan\n",
    "df_assoc[\"AFR_af\"] = np.nan\n",
    "\n",
    "df_assoc[\"EUR_beta\"] = np.nan\n",
    "df_assoc[\"EUR_beta_stderr\"] = np.nan\n",
    "df_assoc[\"AFR_beta\"] = np.nan\n",
    "df_assoc[\"AFR_beta_stderr\"] = np.nan\n",
    "df_assoc[\"HET_pval\"] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8be09bd3-924d-4723-822d-1c694de2faf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "211it [01:20,  2.62it/s]\n"
     ]
    }
   ],
   "source": [
    "for row_i, row in tqdm(df_assoc.iterrows()):\n",
    "    chrom = row.chromosome\n",
    "    position = row.position\n",
    "    if row.id not in dset_gwas_hit.snp.values:\n",
    "        continue\n",
    "    dset_snp = dset_gwas_hit.sel(snp=[row.id])\n",
    "    dset_snp = dset_snp.sel(indiv=~np.isnan(dset_snp[f\"{row.trait}@indiv\"]))\n",
    "    apa = dset_snp[\"allele_per_anc\"].values[:, 0, :]\n",
    "    admix.tools.af_per_anc(dset_snp)\n",
    "    df_covar = dict()\n",
    "    for col in [\"age\", \"sex\", \"study\"] + [f\"geno_EV{i}\" for i in range(1, 11)]:\n",
    "        df_covar[col] = dset_snp[col + \"@indiv\"].values\n",
    "    df_covar = pd.DataFrame(df_covar)\n",
    "    study_dummies = pd.get_dummies(df_covar[\"study\"], drop_first=True)\n",
    "    study_dummies.columns = [f\"study_dummy_{s}\" for s in study_dummies.columns]\n",
    "    df_covar = pd.concat([df_covar, study_dummies], axis=1)\n",
    "    df_covar = df_covar.drop(columns=[\"study\"])\n",
    "    covar = df_covar.values\n",
    "    covar = (covar - covar.mean(axis=0)) / covar.std(axis=0)\n",
    "    y = dset_snp[f\"{row.trait}@indiv\"].values\n",
    "    y = quantile_normalize(y)\n",
    "    p_het, model_het = test_het(apa, y, covar)\n",
    "    p_assoc, model_assoc = test_assoc(apa, y, covar)\n",
    "    df_assoc.loc[row_i, [\"EUR_af\", \"AFR_af\"]] = (\n",
    "        dset_snp[\"af_per_anc\"].values[0, 0],\n",
    "        dset_snp[\"af_per_anc\"].values[0, 1],\n",
    "    )\n",
    "    df_assoc.loc[\n",
    "        row_i,\n",
    "        [\n",
    "            \"assoc_pval\",\n",
    "            \"HET_pval\",\n",
    "            \"EUR_beta\",\n",
    "            \"AFR_beta\",\n",
    "            \"EUR_beta_stderr\",\n",
    "            \"AFR_beta_stderr\",\n",
    "        ],\n",
    "    ] = [\n",
    "        p_assoc,\n",
    "        p_het,\n",
    "        model_het.params[1],\n",
    "        model_het.params[2],\n",
    "        model_het.bse[1],\n",
    "        model_het.bse[2],\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "220c0718-36df-4502-97fd-98a9d0757911",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>id</th>\n",
       "      <th>rsId</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>nearestCodingGene</th>\n",
       "      <th>nearestCodingGeneDistance</th>\n",
       "      <th>GWAS_catalog_pval</th>\n",
       "      <th>ldSetSize</th>\n",
       "      <th>EUR_af</th>\n",
       "      <th>AFR_af</th>\n",
       "      <th>EUR_beta</th>\n",
       "      <th>EUR_beta_stderr</th>\n",
       "      <th>AFR_beta</th>\n",
       "      <th>AFR_beta_stderr</th>\n",
       "      <th>HET_pval</th>\n",
       "      <th>assoc_pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>crp</td>\n",
       "      <td>1_65624099_C_T</td>\n",
       "      <td>rs6700896</td>\n",
       "      <td>1</td>\n",
       "      <td>65624099.0</td>\n",
       "      <td>PDE4B</td>\n",
       "      <td>168415.0</td>\n",
       "      <td>2.000000e-25</td>\n",
       "      <td>129.0</td>\n",
       "      <td>0.408433</td>\n",
       "      <td>0.465942</td>\n",
       "      <td>-0.078825</td>\n",
       "      <td>0.026669</td>\n",
       "      <td>-0.058604</td>\n",
       "      <td>0.016611</td>\n",
       "      <td>0.475105</td>\n",
       "      <td>3.624774e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>crp</td>\n",
       "      <td>1_154454494_A_C</td>\n",
       "      <td>rs2228145</td>\n",
       "      <td>1</td>\n",
       "      <td>154454494.0</td>\n",
       "      <td>TDRD10</td>\n",
       "      <td>47725.0</td>\n",
       "      <td>9.000000e-34</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.412968</td>\n",
       "      <td>0.076964</td>\n",
       "      <td>-0.156980</td>\n",
       "      <td>0.026353</td>\n",
       "      <td>-0.016431</td>\n",
       "      <td>0.032193</td>\n",
       "      <td>0.000445</td>\n",
       "      <td>1.553822e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>crp</td>\n",
       "      <td>1_159685728_G_A</td>\n",
       "      <td>rs726640</td>\n",
       "      <td>1</td>\n",
       "      <td>159685728.0</td>\n",
       "      <td>CRP</td>\n",
       "      <td>28861.0</td>\n",
       "      <td>6.000000e-117</td>\n",
       "      <td>47.0</td>\n",
       "      <td>0.001030</td>\n",
       "      <td>0.208739</td>\n",
       "      <td>0.232544</td>\n",
       "      <td>0.483246</td>\n",
       "      <td>0.365418</td>\n",
       "      <td>0.020218</td>\n",
       "      <td>0.783419</td>\n",
       "      <td>1.096978e-71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>crp</td>\n",
       "      <td>1_247438293_T_C</td>\n",
       "      <td>rs12239046</td>\n",
       "      <td>1</td>\n",
       "      <td>247438293.0</td>\n",
       "      <td>OR2B11</td>\n",
       "      <td>19812.0</td>\n",
       "      <td>1.000000e-09</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.638524</td>\n",
       "      <td>0.451357</td>\n",
       "      <td>0.035252</td>\n",
       "      <td>0.022846</td>\n",
       "      <td>0.035906</td>\n",
       "      <td>0.016719</td>\n",
       "      <td>0.978590</td>\n",
       "      <td>1.791943e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>crp</td>\n",
       "      <td>2_27508073_T_C</td>\n",
       "      <td>rs1260326</td>\n",
       "      <td>2</td>\n",
       "      <td>27508073.0</td>\n",
       "      <td>GCKR</td>\n",
       "      <td>11234.0</td>\n",
       "      <td>3.000000e-27</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.604076</td>\n",
       "      <td>0.918421</td>\n",
       "      <td>-0.094591</td>\n",
       "      <td>0.027497</td>\n",
       "      <td>-0.065589</td>\n",
       "      <td>0.021587</td>\n",
       "      <td>0.212499</td>\n",
       "      <td>5.435033e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>bmi</td>\n",
       "      <td>13_53528071_G_A</td>\n",
       "      <td>rs12429545</td>\n",
       "      <td>13</td>\n",
       "      <td>53528071.0</td>\n",
       "      <td>OLFM4</td>\n",
       "      <td>499258.0</td>\n",
       "      <td>2.000000e-08</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.155075</td>\n",
       "      <td>0.018685</td>\n",
       "      <td>-0.014697</td>\n",
       "      <td>0.029416</td>\n",
       "      <td>0.076984</td>\n",
       "      <td>0.043621</td>\n",
       "      <td>0.079769</td>\n",
       "      <td>5.743546e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>207</th>\n",
       "      <td>bmi</td>\n",
       "      <td>16_3963466_C_T</td>\n",
       "      <td>rs2531995</td>\n",
       "      <td>16</td>\n",
       "      <td>3963466.0</td>\n",
       "      <td>CREBBP</td>\n",
       "      <td>82740.0</td>\n",
       "      <td>5.000000e-09</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.579632</td>\n",
       "      <td>0.094109</td>\n",
       "      <td>0.001340</td>\n",
       "      <td>0.016689</td>\n",
       "      <td>0.041867</td>\n",
       "      <td>0.020054</td>\n",
       "      <td>0.102777</td>\n",
       "      <td>1.909695e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208</th>\n",
       "      <td>bmi</td>\n",
       "      <td>16_53769662_T_A</td>\n",
       "      <td>rs1558902</td>\n",
       "      <td>16</td>\n",
       "      <td>53769662.0</td>\n",
       "      <td>RPGRIP1L</td>\n",
       "      <td>65724.0</td>\n",
       "      <td>4.000000e-29</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.381213</td>\n",
       "      <td>0.038373</td>\n",
       "      <td>0.063977</td>\n",
       "      <td>0.019643</td>\n",
       "      <td>0.055455</td>\n",
       "      <td>0.030748</td>\n",
       "      <td>0.810922</td>\n",
       "      <td>2.757239e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>bmi</td>\n",
       "      <td>18_60161902_T_C</td>\n",
       "      <td>rs6567160</td>\n",
       "      <td>18</td>\n",
       "      <td>60161902.0</td>\n",
       "      <td>MC4R</td>\n",
       "      <td>210873.0</td>\n",
       "      <td>2.000000e-11</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.208939</td>\n",
       "      <td>0.182251</td>\n",
       "      <td>0.086659</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>0.050865</td>\n",
       "      <td>0.015117</td>\n",
       "      <td>0.212765</td>\n",
       "      <td>9.926148e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>egfrckdepi</td>\n",
       "      <td>9_68819791_A_C</td>\n",
       "      <td>rs4744712</td>\n",
       "      <td>9</td>\n",
       "      <td>68819791.0</td>\n",
       "      <td>FAM122A</td>\n",
       "      <td>39757.0</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.626555</td>\n",
       "      <td>0.569231</td>\n",
       "      <td>0.019963</td>\n",
       "      <td>0.021928</td>\n",
       "      <td>0.047679</td>\n",
       "      <td>0.014426</td>\n",
       "      <td>0.212266</td>\n",
       "      <td>2.200867e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>211 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          trait               id        rsId chromosome     position  \\\n",
       "0           crp   1_65624099_C_T   rs6700896          1   65624099.0   \n",
       "1           crp  1_154454494_A_C   rs2228145          1  154454494.0   \n",
       "2           crp  1_159685728_G_A    rs726640          1  159685728.0   \n",
       "3           crp  1_247438293_T_C  rs12239046          1  247438293.0   \n",
       "4           crp   2_27508073_T_C   rs1260326          2   27508073.0   \n",
       "..          ...              ...         ...        ...          ...   \n",
       "206         bmi  13_53528071_G_A  rs12429545         13   53528071.0   \n",
       "207         bmi   16_3963466_C_T   rs2531995         16    3963466.0   \n",
       "208         bmi  16_53769662_T_A   rs1558902         16   53769662.0   \n",
       "209         bmi  18_60161902_T_C   rs6567160         18   60161902.0   \n",
       "210  egfrckdepi   9_68819791_A_C   rs4744712          9   68819791.0   \n",
       "\n",
       "    nearestCodingGene  nearestCodingGeneDistance  GWAS_catalog_pval  \\\n",
       "0               PDE4B                   168415.0       2.000000e-25   \n",
       "1              TDRD10                    47725.0       9.000000e-34   \n",
       "2                 CRP                    28861.0      6.000000e-117   \n",
       "3              OR2B11                    19812.0       1.000000e-09   \n",
       "4                GCKR                    11234.0       3.000000e-27   \n",
       "..                ...                        ...                ...   \n",
       "206             OLFM4                   499258.0       2.000000e-08   \n",
       "207            CREBBP                    82740.0       5.000000e-09   \n",
       "208          RPGRIP1L                    65724.0       4.000000e-29   \n",
       "209              MC4R                   210873.0       2.000000e-11   \n",
       "210           FAM122A                    39757.0       1.000000e-08   \n",
       "\n",
       "     ldSetSize    EUR_af    AFR_af  EUR_beta  EUR_beta_stderr  AFR_beta  \\\n",
       "0        129.0  0.408433  0.465942 -0.078825         0.026669 -0.058604   \n",
       "1         21.0  0.412968  0.076964 -0.156980         0.026353 -0.016431   \n",
       "2         47.0  0.001030  0.208739  0.232544         0.483246  0.365418   \n",
       "3          3.0  0.638524  0.451357  0.035252         0.022846  0.035906   \n",
       "4          5.0  0.604076  0.918421 -0.094591         0.027497 -0.065589   \n",
       "..         ...       ...       ...       ...              ...       ...   \n",
       "206        6.0  0.155075  0.018685 -0.014697         0.029416  0.076984   \n",
       "207        8.0  0.579632  0.094109  0.001340         0.016689  0.041867   \n",
       "208       29.0  0.381213  0.038373  0.063977         0.019643  0.055455   \n",
       "209       17.0  0.208939  0.182251  0.086659         0.025641  0.050865   \n",
       "210       17.0  0.626555  0.569231  0.019963         0.021928  0.047679   \n",
       "\n",
       "     AFR_beta_stderr  HET_pval    assoc_pval  \n",
       "0           0.016611  0.475105  3.624774e-05  \n",
       "1           0.032193  0.000445  1.553822e-06  \n",
       "2           0.020218  0.783419  1.096978e-71  \n",
       "3           0.016719  0.978590  1.791943e-02  \n",
       "4           0.021587  0.212499  5.435033e-04  \n",
       "..               ...       ...           ...  \n",
       "206         0.043621  0.079769  5.743546e-01  \n",
       "207         0.020054  0.102777  1.909695e-01  \n",
       "208         0.030748  0.810922  2.757239e-04  \n",
       "209         0.015117  0.212765  9.926148e-06  \n",
       "210         0.014426  0.212266  2.200867e-03  \n",
       "\n",
       "[211 rows x 17 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_assoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "id": "a9c898b3-24dc-4dab-b371-cecb8782a70d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_assoc.to_excel(\"out/locus_het/sumstats.xlsx\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
