{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "051181f6-c6da-4461-8b0b-6684170c1fc1",
   "metadata": {},
   "source": [
    "# Simulation to evaluate the reason of inflation\n",
    "\n",
    "1. Take the chromosome-wide simulation in 1st experiment.\n",
    "2. Perform a GWAS on the data on chromosome 22.\n",
    "3. Perform clumping on the data on chromosome 22.\n",
    "4. For every GWAS hit, find the true SNPs, and evaluate the properties. Heterogeneity ~ distance to causal, strategies to find causal variables.\n",
    "5. Run heterogeneity test on the index SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4d6a2af3-09e0-4add-a5a9-60d27be0a702",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import json\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import os\n",
    "import admix\n",
    "from os.path import join\n",
    "import itertools\n",
    "import admix_genet_cor\n",
    "import submitit\n",
    "import glob\n",
    "from scipy.stats import pearsonr, linregress\n",
    "from tqdm import tqdm\n",
    "import string\n",
    "from natsort import natsorted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82a8e72e-c58a-4475-9c95-77e63ea27ba1",
   "metadata": {},
   "source": [
    "# Convert PLINK2 to PLINK1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "20e78ac8-3b6e-425c-ac89-c07ec505757a",
   "metadata": {},
   "outputs": [],
   "source": [
    "GENO_DIR = (\n",
    "    \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01-dataset/out/aframr\"\n",
    ")\n",
    "SAMPLE_INFO_PATH = \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01-dataset/out/aframr/sample_info.txt\"\n",
    "\n",
    "# convert to PLINK1\n",
    "def convert_plink1():\n",
    "    for chrom in range(1, 23):\n",
    "        pfile = join(GENO_DIR, f\"imputed/chr{chrom}\")\n",
    "        admix.tools.plink2.run(f\"--pfile {pfile} --make-bed --out out/PLINK/chr{chrom}\")\n",
    "\n",
    "\n",
    "# convert_plink1()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92d0bbc2-4527-4f59-9e69-982441ad4b3c",
   "metadata": {},
   "source": [
    "# Perform GWAS clumping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0f77f890-515f-4a5d-a223-e39f8590ade1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gwas_clump(config, sim_i):\n",
    "\n",
    "    dset_chrom = admix.io.read_dataset(\n",
    "        pfile=join(GENO_DIR, f\"imputed/chr1\"),\n",
    "        indiv_info_file=SAMPLE_INFO_PATH,\n",
    "        n_anc=2,\n",
    "    )\n",
    "\n",
    "    df_sample_info = pd.read_csv(\n",
    "        f\"out/pheno/{config}.pheno.tsv.gz\", sep=\"\\t\", index_col=0\n",
    "    )\n",
    "    # extract covariates and perform standardization\n",
    "    df_cov = dset_chrom.indiv[[\"geno_EV1\"]]\n",
    "    df_cov = (df_cov - df_cov.mean(axis=0)) / df_cov.std(axis=0)\n",
    "    df_sample_info = pd.merge(\n",
    "        df_sample_info[[f\"SIM_{sim_i}\"]],\n",
    "        df_cov,\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "    )\n",
    "\n",
    "    for chrom in range(1, 2):\n",
    "        out_prefix = f\"out/clump/{config}.sim_{sim_i}.chr{chrom}\"\n",
    "        admix.tools.plink2.gwas(\n",
    "            pfile=join(GENO_DIR, f\"imputed/chr{chrom}\"),\n",
    "            df_sample_info=df_sample_info,\n",
    "            pheno_col=f\"SIM_{sim_i}\",\n",
    "            covar_cols=[\"geno_EV1\"],\n",
    "            out_prefix=out_prefix,\n",
    "            clean_tmp_file=True,\n",
    "            memory=int(28 * 1e3),\n",
    "        )\n",
    "        # See Pardinas et al. for description of these set of parameters\n",
    "        admix.tools.plink.clump(\n",
    "            bfile=f\"out/PLINK/chr{chrom}\",\n",
    "            assoc_path=out_prefix + \".assoc\",\n",
    "            out_prefix=out_prefix,\n",
    "            p1=5e-8,\n",
    "            p2=1e-4,\n",
    "            r2=0.1,\n",
    "            kb=10000,  # 10Mb clumping window\n",
    "            memory=int(28 * 1e3),\n",
    "        )\n",
    "        os.remove(out_prefix + \".assoc\")\n",
    "\n",
    "    # merging all the clumping\n",
    "    df_clump = []\n",
    "    clump_file_prefix = f\"out/clump/{config}.sim_{sim_i}\"\n",
    "    for chrom in range(1, 2):\n",
    "        clumped_file = clump_file_prefix + f\".chr{chrom}.clumped\"\n",
    "        with open(clumped_file) as f:\n",
    "            lines = f.readlines()\n",
    "            if len(lines) == 1:\n",
    "                continue\n",
    "        df_clump.append(pd.read_csv(clumped_file, delim_whitespace=True))\n",
    "    df_clump = pd.concat(df_clump).sort_values([\"CHR\", \"BP\"])\n",
    "    df_clump.to_csv(clump_file_prefix + f\".clumped\", index=False, sep=\"\\t\")\n",
    "    # clean up\n",
    "    for f in glob.glob(clump_file_prefix + \".chr*\"):\n",
    "        os.remove(f)\n",
    "\n",
    "\n",
    "def test_het(config, sim_i):\n",
    "    ###########\n",
    "    # read info\n",
    "    ###########\n",
    "    df_clump = (\n",
    "        pd.read_csv(f\"out/clump/{config}.sim_{sim_i}.clumped\", sep=\"\\t\")\n",
    "        .set_index(\"SNP\")\n",
    "        .sort_values([\"CHR\", \"BP\"])\n",
    "    )\n",
    "    df_pheno = pd.read_csv(f\"out/pheno/{config}.pheno.tsv.gz\", sep=\"\\t\", index_col=0)[\n",
    "        f\"SIM_{sim_i}\"\n",
    "    ]\n",
    "\n",
    "    df_beta = pd.read_csv(f\"out/pheno/{config}.beta_info.tsv.gz\", sep=\"\\t\", index_col=0)\n",
    "    # attached effect sizes\n",
    "    df_beta[[\"AFR_BETA\", \"EUR_BETA\"]] = np.load(f\"out/pheno/{config}.beta.npz\",)[\n",
    "        \"arr_0\"\n",
    "    ][:, :, sim_i]\n",
    "    assert np.allclose(df_beta[\"AFR_BETA\"], df_beta[\"EUR_BETA\"])\n",
    "\n",
    "    # extract only causal SNPs\n",
    "    # TODO: maybe threhold this such that the expected p-value < 5e-8\n",
    "    dict_snp_list = {\n",
    "        \"causal\": df_beta[df_beta.AFR_BETA != 0].index.values,\n",
    "        \"clump\": df_clump.index.values,\n",
    "    }\n",
    "\n",
    "    dict_df_summary = {\"clump\": [], \"causal\": []}\n",
    "\n",
    "    for chrom in range(1, 2):\n",
    "        dset_chrom = admix.io.read_dataset(\n",
    "            pfile=join(GENO_DIR, f\"imputed/chr{chrom}\"),\n",
    "            indiv_info_file=SAMPLE_INFO_PATH,\n",
    "            n_anc=2,\n",
    "        )\n",
    "        for group in [\"clump\", \"causal\"]:\n",
    "            snp_list = [\n",
    "                snp for snp in dict_snp_list[group] if snp.startswith(f\"chr{chrom}:\")\n",
    "            ]\n",
    "            if len(snp_list) == 0:\n",
    "                continue\n",
    "            dset_tmp = dset_chrom[snp_list]\n",
    "            apa = dset_tmp.allele_per_anc()\n",
    "\n",
    "            cov_values = dset_chrom.indiv[[\"geno_EV1\"]].values\n",
    "            cov_values = (cov_values - cov_values.mean(axis=0)) / cov_values.std(axis=0)\n",
    "            # heterogeneity test\n",
    "            df_tmp = admix_genet_cor.marginal_het(\n",
    "                geno=dset_tmp.geno,\n",
    "                lanc=dset_tmp.lanc,\n",
    "                y=df_pheno,\n",
    "                cov=cov_values,\n",
    "            )\n",
    "            # association test\n",
    "            df_tmp[\"assoc_p\"] = admix.assoc.marginal(\n",
    "                dset=dset_tmp,\n",
    "                pheno=df_pheno,\n",
    "                cov=cov_values,\n",
    "                method=\"ATT\",\n",
    "            ).P.values\n",
    "            df_tmp[\"snp\"] = dset_tmp.snp.index.values\n",
    "            dict_df_summary[group].append(df_tmp)\n",
    "\n",
    "    for group in [\"clump\", \"causal\"]:\n",
    "        df_tmp = pd.concat(dict_df_summary[group])\n",
    "        df_tmp = df_tmp[\n",
    "            [\n",
    "                \"snp\",\n",
    "                \"het_pval\",\n",
    "                \"coef1\",\n",
    "                \"se1\",\n",
    "                \"coef2\",\n",
    "                \"se2\",\n",
    "                \"assoc_p\",\n",
    "            ]\n",
    "        ]\n",
    "        df_tmp.to_csv(\n",
    "            f\"out/summary/{config}.sim_{sim_i}.{group}.tsv\",\n",
    "            index=False,\n",
    "            sep=\"\\t\",\n",
    "        )\n",
    "\n",
    "\n",
    "def submit_job(config, sim_i):\n",
    "    gwas_clump(config, sim_i)\n",
    "    test_het(config, sim_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "752d36c8-2b4b-4d1e-8f18-414240579634",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define the simulation parameters\n",
    "df_params = pd.DataFrame(\n",
    "    [\n",
    "        params\n",
    "        for params in itertools.product(\n",
    "            [0.025, 0.05, 0.1, 0.2], [62, 125, 250, 500, 1000], np.arange(0, 500)\n",
    "        )\n",
    "    ],\n",
    "    columns=[\"hsq\", \"ncausal\", \"sim_i\"],\n",
    ")\n",
    "df_params[\"prefix\"] = df_params.apply(\n",
    "    lambda row: f\"hsq-{row.hsq}-ncausal-{int(row.ncausal)}\",\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "df_todo_params = df_params[\n",
    "    ~df_params.apply(\n",
    "        lambda x: os.path.exists(\n",
    "            join(\"out/summary\", x.prefix + f\".sim_{x.sim_i}.causal.tsv\")\n",
    "        ),\n",
    "        axis=1,\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ea5062d9-24b7-48f4-a323-62f9f94e9068",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=180,\n",
    "    memory_g=36,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(submit_job, df_todo_params.prefix, df_todo_params.sim_i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c9e2b9e8-43c7-4d89-ba44-2a7985d8996c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hsq-0.025-ncausal-1000    435\n",
       "hsq-0.025-ncausal-500     376\n",
       "hsq-0.05-ncausal-1000     247\n",
       "hsq-0.025-ncausal-250     224\n",
       "hsq-0.05-ncausal-500       71\n",
       "hsq-0.025-ncausal-125      36\n",
       "hsq-0.1-ncausal-1000        5\n",
       "hsq-0.05-ncausal-250        2\n",
       "hsq-0.025-ncausal-62        1\n",
       "Name: prefix, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_todo_params[\"prefix\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e67f9f4-0beb-47f8-bccb-f5fd0315950f",
   "metadata": {},
   "source": [
    "# Legacy code below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3fe1ea3e-9c52-4828-b82e-e40b3c5d31b0",
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-a871fdc9ebee>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "assert False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19fb9ce-b27a-4aa2-8407-768505c0f167",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sum = []\n",
    "for chrom, df_chrom in df_clump.groupby(\"CHR\"):\n",
    "\n",
    "    df_tmp_snp = {\"snp\": [], \"index_snp\": [], \"is_causal\": [], \"clump_size\": []}\n",
    "\n",
    "    for index_snp, row in df_chrom.iterrows():\n",
    "        if row.SP2 != \"NONE\":\n",
    "            snp_list = [s[:-3] for s in row.SP2.split(\",\")]\n",
    "            clump_size = len(snp_list)\n",
    "            # find tag SNPs that are also causal snp\n",
    "            snp_list = [s for s in snp_list if s in causal_snp_list]\n",
    "        else:\n",
    "            snp_list = []\n",
    "            clump_size = 1\n",
    "        snp_list.append(index_snp)\n",
    "        df_tmp_snp[\"snp\"].extend(snp_list)\n",
    "        df_tmp_snp[\"index_snp\"].extend([index_snp] * len(snp_list))\n",
    "        df_tmp_snp[\"is_causal\"].extend([False] + [True] * (len(snp_list) - 1))\n",
    "        df_tmp_snp[\"clump_size\"].extend([clump_size] * len(snp_list))\n",
    "\n",
    "    df_tmp_snp = pd.DataFrame(df_tmp_snp).set_index(\"snp\")\n",
    "    df_tmp_snp = df_tmp_snp.loc[natsorted(df_tmp_snp.index)]\n",
    "\n",
    "    dset_chrom = admix.io.read_dataset(\n",
    "        pfile=join(GENO_DIR, f\"imputed/chr{chrom}\"), n_anc=2\n",
    "    )[df_tmp_snp.index.values]\n",
    "\n",
    "    df_tmp_corr = admix_genet_cor.marginal_het(\n",
    "        dset_chrom.geno, dset_chrom.lanc, df_pheno\n",
    "    )\n",
    "    df_tmp_corr[\"assoc_p\"] = admix.assoc.marginal(\n",
    "        geno=dset_chrom.geno, lanc=dset_chrom.lanc, pheno=df_pheno, cov=None\n",
    "    )\n",
    "    df_tmp_corr.index = df_tmp_snp.index.values"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
