{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4c71feda-724a-49b0-96e9-ad0a7075d76f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "\n",
    "import numpy as np\n",
    "import dapgen\n",
    "import pandas as pd\n",
    "import dask.array as da\n",
    "import itertools\n",
    "import submitit\n",
    "import json\n",
    "import matplotlib.pyplot as plt\n",
    "import admix_genet_cor\n",
    "from admix_genet_cor import calc_snp_prior_var\n",
    "import pandas as pd\n",
    "import admix\n",
    "from os.path import join\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afc8e62e-5355-40f4-b76f-017d25b8cba7",
   "metadata": {},
   "source": [
    "# Overview of simulation studies\n",
    "- p_causal: 0.001, 1.0\n",
    "- var_g: 1.0\n",
    "- var_e: 1.0\n",
    "- gamma: 0.5, 0.8, 1.0\n",
    "- set of SNPs: simulate from the imputed SNPs, the non-zero effects are simulated from SNPs with allele frequencies > 0.005 in both populations.\n",
    "\n",
    "From S-LDXR paper: \n",
    "```\n",
    "In these simulations, we randomly selected 10% of the SNPs to be causal in each population, with 80% of causal variants in each population shared with the other population, and sampled perfectly correlated causal effect sizes for shared causal variants using Eq. In these simulations, we set the variance of causal effect size of each SNP j in both populations to be proportional to $[p_{j,max}(1−p_{j,max})]^\\alpha$, where pj,max is the maximum MAF of SNP j in the two populations.  We set α to − 0.38, as previously estimated for 25 UK Biobank diseases \n",
    "and complex traits in ref.\n",
    "```\n",
    "\n",
    "Here, we simulate from the imputed SNPs the non-zero effects are simulated from SNPs with allele frequencies > 0.005 in both populations. And from these, we sample 1% of causal effects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b78dd1f9-7e8d-4c0f-ad57-553122ea893b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CONSTANTS\n",
    "\n",
    "DATA_ROOT_DIR = (\n",
    "    \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01_dataset/out/aframr\"\n",
    ")\n",
    "GRM_DIR = \"/u/project/pasaniuc/kangchen/2021-admix-corr/experiments/03_page_genome_wide/out/admix-grm\"\n",
    "\n",
    "# define the simulation parameters\n",
    "df_simulate_params = pd.DataFrame(\n",
    "    [\n",
    "        params\n",
    "        for params in itertools.product(\n",
    "            [0.1, 0.25, 0.5],\n",
    "            [0.00001, 0.0001, 0.001, 0.01],\n",
    "            [0.9, 0.95, 1.0],\n",
    "            [\"mafukb\", \"uniform\", \"gcta\"],\n",
    "        )\n",
    "    ],\n",
    "    columns=[\"hsq\", \"pcausal\", \"cor\", \"hermodel\"],\n",
    ")\n",
    "df_simulate_params[\"out_prefix\"] = df_simulate_params.apply(\n",
    "    lambda row: f\"out/pheno/hsq-{row.hsq}-pcausal-{row.pcausal}-cor-{row.cor}-hermodel-{row.hermodel}\",\n",
    "    axis=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b23a3f5-4ebc-4b32-9cbf-e6b9dec8a060",
   "metadata": {},
   "source": [
    "# Step 1: Simulate phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "54e09928-ccda-4f45-9749-c0d06d72cf81",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_simulate_pheno(\n",
    "    hsq: float,\n",
    "    pcausal: float,\n",
    "    cor: float,\n",
    "    hermodel: str,\n",
    "    out_prefix: str,\n",
    "    n_sim=100,\n",
    "):\n",
    "    \"\"\"\n",
    "    her_model: one of [uniform, gcta, ldak]\n",
    "\n",
    "    \"\"\"\n",
    "    np.random.seed(admix.utils.str2int(out_prefix))\n",
    "    N_ANC = 2\n",
    "    pfile_list = [f\"{DATA_ROOT_DIR}/imputed/chr{chrom}\" for chrom in range(1, 23)]\n",
    "\n",
    "    geno = []\n",
    "    lanc = []\n",
    "    df_indiv = None\n",
    "    df_snp = []\n",
    "    df_snp_info = []\n",
    "\n",
    "    # read data\n",
    "    for pfile in pfile_list:\n",
    "\n",
    "        this_geno, this_df_snp, this_df_indiv = dapgen.read_pfile(\n",
    "            pfile, phase=True, snp_chunk=2048\n",
    "        )\n",
    "        this_lanc = admix.io.read_lanc(pfile + \".lanc\").dask(snp_chunk=2048)\n",
    "        this_df_snp_info = pd.read_csv(pfile + \".snp_info\", sep=\"\\t\")\n",
    "        assert np.all(this_df_snp_info.SNP == this_df_snp.index.values)\n",
    "\n",
    "        if df_indiv is None:\n",
    "            df_indiv = this_df_indiv\n",
    "        else:\n",
    "            assert df_indiv.equals(\n",
    "                df_indiv\n",
    "            ), \".psam should be consistent for all pfiles\"\n",
    "        geno.append(this_geno)\n",
    "        lanc.append(this_lanc)\n",
    "        df_snp_info.append(this_df_snp_info)\n",
    "\n",
    "    # concatenate\n",
    "    geno = da.concatenate(geno, axis=0)\n",
    "    lanc = da.concatenate(lanc, axis=0)\n",
    "    df_snp_info = pd.concat(df_snp_info).reset_index(drop=True)\n",
    "\n",
    "    snp_prior_var = calc_snp_prior_var(df_snp_info, hermodel)\n",
    "\n",
    "    # simulate effects\n",
    "    snp_subset = np.where(\n",
    "        (df_snp_info.EUR_FREQ.between(0.005, 0.995) & (df_snp_info.AFR_FREQ.between(0.005, 0.995))\n",
    "    )[0]\n",
    "\n",
    "    # sub-sample SNPs from `geno`, `lanc`, `df_snp_info`, `snp_prior_var`\n",
    "    n_eff_snp = len(snp_subset)\n",
    "    geno = geno[snp_subset, :, :]\n",
    "    lanc = lanc[snp_subset, :, :]\n",
    "    df_snp_info = df_snp_info.iloc[snp_subset, :]\n",
    "    snp_prior_var = snp_prior_var[snp_subset]\n",
    "\n",
    "    beta = np.zeros((n_eff_snp, N_ANC, n_sim))  # (n_snp, n_anc, n_sim)\n",
    "    n_causal = int(n_eff_snp * pcausal)\n",
    "    print(f\"n_causal: {n_causal}\")\n",
    "    for i_sim in range(n_sim):\n",
    "        cau = sorted(\n",
    "            np.random.choice(np.arange(n_eff_snp), size=n_causal, replace=False)\n",
    "        )\n",
    "\n",
    "        i_beta = np.random.multivariate_normal(\n",
    "            mean=[0.0, 0.0],\n",
    "            cov=np.array([[1, cor], [cor, 1]]) / n_causal,\n",
    "            size=n_causal,\n",
    "        )\n",
    "\n",
    "        i_beta = i_beta * np.sqrt(snp_prior_var[cau])[:, None]\n",
    "\n",
    "        for i_anc in range(N_ANC):\n",
    "            beta[cau, i_anc, i_sim] = i_beta[:, i_anc]\n",
    "\n",
    "    sim = admix_genet_cor.simulate_quant_pheno(\n",
    "        geno=geno, lanc=lanc, hsq=hsq, beta=beta, n_sim=n_sim\n",
    "    )\n",
    "    np.savez_compressed(out_prefix + \".beta\", sim[\"beta\"])\n",
    "    df_snp_info.to_csv(out_prefix + \".beta_info.tsv.gz\", index=False, sep=\"\\t\")\n",
    "\n",
    "    df_pheno = pd.DataFrame(\n",
    "        sim[\"pheno\"],\n",
    "        index=df_indiv.index,\n",
    "        columns=[f\"SIM_{i}\" for i in range(n_sim)],\n",
    "    )\n",
    "    df_pheno.to_csv(out_prefix + \".pheno.tsv.gz\", index=True, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "30403662-46f3-4fe6-90d0-82d417f244d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_todo_params = df_simulate_params[\n",
    "    ~df_simulate_params.apply(\n",
    "        lambda x: os.path.exists(x.out_prefix + \".pheno.tsv.gz\"), axis=1\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "27a2381d-6e93-4073-afe3-ad3c07ae3139",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=1200,\n",
    "    memory_g=30,\n",
    "    queue=\"highp\",\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "df_todo_params = df_simulate_params[\n",
    "    ~df_simulate_params.apply(\n",
    "        lambda x: os.path.exists(x.out_prefix + \".pheno.tsv.gz\"), axis=1\n",
    "    )\n",
    "]\n",
    "jobs = executor.map_array(\n",
    "    submit_simulate_pheno,\n",
    "    df_todo_params.hsq,\n",
    "    df_todo_params.pcausal,\n",
    "    df_todo_params.cor,\n",
    "    df_todo_params.hermodel,\n",
    "    df_todo_params.out_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24fadf06-d0bd-497e-9870-5923706bcf7a",
   "metadata": {},
   "source": [
    "# Step 2: GCTA Estimation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dc8326eb-05cb-439c-b008-224a74eb051c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_estimate_params = df_simulate_params.copy()\n",
    "df_estimate_params[\"estimate_out_prefix\"] = df_estimate_params.apply(\n",
    "    lambda row: row.out_prefix.replace(\"pheno\", \"gcta-estimate\"), axis=1\n",
    ")\n",
    "df_estimate_params = df_estimate_params.merge(\n",
    "    pd.DataFrame({\"sim_i\": np.arange(100)}), how=\"cross\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c7b8fb16-bd8b-469a-bbf0-6dc7b6270cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_gcta_estimate(hermodel: str, pheno: str, sim_i: int, out_prefix: str):\n",
    "    # compile phenotype and covariates\n",
    "    print(f\"submit_estimate:\\npheno={pheno}, out_prefix={out_prefix}\")\n",
    "\n",
    "    df_pheno = pd.read_csv(pheno, delim_whitespace=True, index_col=0)\n",
    "    df_pheno = pd.DataFrame(\n",
    "        {\n",
    "            \"FID\": df_pheno.index.values,\n",
    "            \"IID\": df_pheno.index.values,\n",
    "            \"trait\": df_pheno[f\"SIM_{sim_i}\"].values,\n",
    "        }\n",
    "    )\n",
    "    # fit reduced model\n",
    "    reduced_grm = join(GRM_DIR, f\"imputed.{hermodel}.all.K_full\")\n",
    "    admix.tools.gcta.reml(\n",
    "        grm_path=reduced_grm,\n",
    "        df_pheno=df_pheno,\n",
    "        out_prefix=out_prefix + f\".sim_{sim_i}.reduced\",\n",
    "    )\n",
    "    # fit full model\n",
    "    mgrm_path = join(GRM_DIR, f\"imputed.{hermodel}.all.mgrm.txt\")\n",
    "    admix.tools.gcta.reml(\n",
    "        mgrm_path=mgrm_path, df_pheno=df_pheno, out_prefix=out_prefix + f\".sim_{sim_i}\"\n",
    "    )\n",
    "    # remove tmp pheno files\n",
    "    os.remove(out_prefix + f\".sim_{sim_i}.reduced.pheno\")\n",
    "    os.remove(out_prefix + f\".sim_{sim_i}.pheno\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8d300e26-f613-4816-83f4-6b5abefd9b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=50,\n",
    "    memory_g=24,\n",
    "    queue=\"highp\",\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_gcta_estimate,\n",
    "    df_estimate_params.hermodel,\n",
    "    (df_estimate_params[\"out_prefix\"] + \".pheno.tsv.gz\"),\n",
    "    df_estimate_params.sim_i,\n",
    "    df_estimate_params.estimate_out_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f7d2657-4eea-4005-b138-b08cc5f19897",
   "metadata": {},
   "source": [
    "# HE regression\n",
    "Seems HE regression is not as robust as REML. So we didn't bother to use HE regression here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "511d4750-6be7-4607-966b-a3245e754d89",
   "metadata": {},
   "outputs": [
    {
     "ename": "AssertionError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-a871fdc9ebee>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "assert False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e49614c8-c2cf-4fdb-b5a5-a228898a7a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_estimate_params = df_simulate_params.copy()\n",
    "df_estimate_params[\"estimate_out_prefix\"] = df_estimate_params.apply(\n",
    "    lambda row: row.out_prefix.replace(\"pheno\", \"estimate\"), axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2684703d-fd74-4d4f-aa98-272d885119bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_estimate(hermodel: str, pheno: str, out_prefix: str):\n",
    "    \"\"\"\n",
    "    snp_set: set of SNPs used as GRM, for estimation\n",
    "    pheno: phenotype file\n",
    "    out_prefix: prefix to the output\n",
    "    \"\"\"\n",
    "    print(f\"submit_estimate:\\npheno={pheno}, out_prefix={out_prefix}\")\n",
    "    # Step 1: load GRM\n",
    "    GRM_DIR = \"/u/project/pasaniuc/kangchen/2021-admix-corr/experiments/03_page_genome_wide/out/admix-grm\"\n",
    "    grm_prefix = join(GRM_DIR, f\"imputed.{hermodel}.chr1\")\n",
    "    A1 = np.load(grm_prefix + \".A1.npy\")\n",
    "    A2 = np.load(grm_prefix + \".A2.npy\")\n",
    "\n",
    "    # Step 2: load phenotype and estimate\n",
    "    pheno = pd.read_csv(pheno, delim_whitespace=True, index_col=0)\n",
    "    print(pheno.head())\n",
    "    n_indiv = pheno.shape[0]\n",
    "\n",
    "    rls_list = admix_genet_cor.estimate_genetic_cor(\n",
    "        A1, A2, pheno.values, cov=np.ones((n_indiv, 1))\n",
    "    )\n",
    "\n",
    "    if len(rls_list[0]) == 2:\n",
    "        data = [\n",
    "            {\"estimate\": rls[0].tolist(), \"varcov\": rls[1].tolist()} for rls in rls_list\n",
    "        ]\n",
    "    else:\n",
    "        data = [{\"estimate\": rls.tolist()} for rls in rls_list]\n",
    "\n",
    "    with open(out_prefix + \".estimate.json\", \"w\") as out:\n",
    "        json.dump(data, out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1204590c-6f2f-4d71-9675-07098239eba1",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=30,\n",
    "    memory_g=36,\n",
    "    queue=\"highp\",\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_estimate,\n",
    "    df_estimate_params.hermodel,\n",
    "    (df_estimate_params[\"out_prefix\"] + \".pheno.tsv.gz\"),\n",
    "    df_estimate_params.estimate_out_prefix,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a46e4ddf-d0cb-405b-9e7b-0b5765b9d8e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def submit_compute_grm(hermodel: str, out_prefix: str):\n",
    "\n",
    "#     pfile = f\"{DATA_ROOT_DIR}/imputed/chr{CHROM}\"\n",
    "\n",
    "#     geno, df_snp, df_indiv = dapgen.read_pfile(pfile, phase=True, snp_chunk=512)\n",
    "#     lanc = admix.io.read_lanc(pfile + \".lanc\", snp_chunk=512)\n",
    "\n",
    "#     df_snp_info = pd.read_csv(pfile + \".snp_info\", sep=\"\\t\")\n",
    "#     snp_prior_var = calc_snp_prior_var(df_snp_info, hermodel)\n",
    "#     print(\"snp_prior_var:\")\n",
    "#     print(snp_prior_var)\n",
    "\n",
    "#     snp_subset = np.where(\n",
    "#         (df_snp_info.EUR_FREQ > 0.005) & (df_snp_info.AFR_FREQ > 0.005)\n",
    "#     )[0]\n",
    "\n",
    "#     K1, K2, K12 = admix_genet_cor.compute_grm(\n",
    "#         geno=geno[snp_subset, :, :],\n",
    "#         lanc=lanc[snp_subset, :, :],\n",
    "#         snp_prior_var=snp_prior_var[snp_subset],\n",
    "#         apa_center=False,\n",
    "#     )\n",
    "\n",
    "#     np.save(out_prefix + \".K1.npy\", K1)\n",
    "#     np.save(out_prefix + \".K2.npy\", K2)\n",
    "#     np.save(out_prefix + \".K12.npy\", K12)\n",
    "    \n",
    "# executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "# executor.update_parameters(\n",
    "#     time_min=180,\n",
    "#     memory_g=40,\n",
    "#     setup=[\n",
    "#         \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "#         \"export PYTHONNOUSERSITE=True\",\n",
    "#     ],\n",
    "# )\n",
    "\n",
    "# df_grm_params = pd.DataFrame(\n",
    "#     [\n",
    "#         params\n",
    "#         for params in itertools.product(\n",
    "#             [\"mafukb\"],\n",
    "#         )\n",
    "#     ],\n",
    "#     columns=[\"hermodel\"],\n",
    "# )\n",
    "\n",
    "# df_grm_params[\"out_prefix\"] = df_grm_params.apply(\n",
    "#     lambda row: f\"out/grm/hermodel-{row.hermodel}\",\n",
    "#     axis=1,\n",
    "# )\n",
    "# jobs = executor.map_array(\n",
    "#     submit_compute_grm,\n",
    "#     df_grm_params.hermodel,\n",
    "#     df_grm_params.out_prefix,\n",
    "# )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
