{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b9a3a36-9d68-463b-b58a-cb58a5932b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import admix\n",
    "import dask\n",
    "import dask.array as da\n",
    "from tqdm import tqdm\n",
    "import statsmodels.api as sm\n",
    "import matplotlib.pyplot as plt\n",
    "from admix.data import quantile_normalize\n",
    "import submitit\n",
    "\n",
    "import admix_genet_cor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b713c7e7-1fc6-4f7e-9f37-9903688f66e7",
   "metadata": {},
   "source": [
    "# Process GWAS hits from PAGE study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d32e6ce5-efff-4813-85f6-2430f675ae06",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPP_TABLE_URL = \"https://www.dropbox.com/s/jck2mhjby2ur55j/supp_tables.xlsx?dl=1\"\n",
    "trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"trait-info\")\n",
    "trait_list = trait_info[\"trait\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2c6cd289-12f0-474c-b69c-8dfec4bcb78b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_assoc = pd.read_csv(\"out/page-gwas-formatted.tsv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5d6cd2dc-5ca1-4894-be95-39db786f80f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dapgen\n",
    "from os.path import join\n",
    "\n",
    "ROOT_DIR = (\n",
    "    \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01-dataset/out/aframr/\"\n",
    ")\n",
    "PFILE_DIR = join(ROOT_DIR, \"imputed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "96447c4c-a6af-471a-8ed7-c7a05a6295a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_locus_hetero(trait, out, duffy_covar=False):\n",
    "    df_assoc_trait = df_assoc[df_assoc.trait_id == trait].copy()\n",
    "\n",
    "    # format phenotype and covariates\n",
    "\n",
    "    df_trait = pd.read_csv(join(ROOT_DIR, f\"pheno/{trait}.tsv\"), sep=\"\\t\", index_col=0)\n",
    "    df_trait.index = df_trait.index.astype(str)\n",
    "    dset = admix.io.read_dataset(\n",
    "        join(PFILE_DIR, \"chr1\"),\n",
    "        n_anc=2,\n",
    "    )\n",
    "    # subset for individuals with non-nan value in df_trait\n",
    "    dset = dset[:, dset.indiv.index.isin(df_trait.index)]\n",
    "    dset.append_indiv_info(df_trait)\n",
    "    covar_cols = df_trait.columns[1:]\n",
    "\n",
    "    df_pheno = dset.indiv[[trait]].copy()\n",
    "    df_covar = dset.indiv[covar_cols].copy()\n",
    "    df_covar = admix.data.convert_dummy(df_covar)\n",
    "\n",
    "    # special case for duffy SNPs, include the duffy SNPs in the covariate\n",
    "    if duffy_covar:\n",
    "        # find closest SNPs to Duffy SNP (GRCH38: 1:159204893)\n",
    "        duffy_snp_loc = np.argmin(np.abs(dset.snp.POS - 159204893))\n",
    "        assert dset.snp.CHROM.iloc[duffy_snp_loc] == 1\n",
    "        duffy_lanc = dset[duffy_snp_loc].lanc.sum(axis=[0, 2]).compute()\n",
    "        df_covar[\"duffy_lanc\"] = duffy_lanc\n",
    "\n",
    "    for col in df_pheno.columns:\n",
    "        df_pheno[col] = admix.data.quantile_normalize(df_pheno[col])\n",
    "\n",
    "    for col in df_covar.columns:\n",
    "        df_covar[col] = admix.data.quantile_normalize(df_covar[col])\n",
    "\n",
    "    # fill na with column mean\n",
    "    df_covar.fillna(df_covar.mean(), inplace=True)\n",
    "\n",
    "    for chrom in range(1, 23):\n",
    "        # subset df_assoc\n",
    "        df_assoc_trait_chrom = df_assoc_trait[df_assoc_trait.CHROM == chrom]\n",
    "\n",
    "        # subset dset\n",
    "        dset = admix.io.read_dataset(\n",
    "            join(PFILE_DIR, f\"chr{chrom}\"),\n",
    "            n_anc=2,\n",
    "        )\n",
    "        dset = dset[\n",
    "            dset.snp.POS.isin(df_assoc_trait_chrom.POS.values).values,\n",
    "            df_pheno.index.values,\n",
    "        ]\n",
    "        dset.persist()\n",
    "\n",
    "        apa = dset.allele_per_anc()\n",
    "        af = dset.af_per_anc()\n",
    "\n",
    "        for snp_i in tqdm(range(dset.n_snp)):\n",
    "            p_het, model_het = admix_genet_cor.test_snp_het(\n",
    "                apa[snp_i, :, :], df_pheno.values, df_covar.values\n",
    "            )\n",
    "            p_assoc, model_assoc = admix_genet_cor.test_snp_assoc(\n",
    "                apa[snp_i, :, :], df_pheno.values, df_covar.values\n",
    "            )\n",
    "            snp_idx = f\"{trait}:{dset.snp.CHROM[snp_i]}:{dset.snp.POS[snp_i]}\"\n",
    "            df_assoc_trait.loc[snp_idx, [\"EUR_af\", \"AFR_af\"]] = (\n",
    "                af[snp_i, 0],\n",
    "                af[snp_i, 1],\n",
    "            )\n",
    "\n",
    "            df_assoc_trait.loc[\n",
    "                snp_idx,\n",
    "                [\n",
    "                    \"assoc_pval\",\n",
    "                    \"HET_pval\",\n",
    "                    \"EUR_beta\",\n",
    "                    \"AFR_beta\",\n",
    "                    \"EUR_beta_stderr\",\n",
    "                    \"AFR_beta_stderr\",\n",
    "                ],\n",
    "            ] = [\n",
    "                p_assoc,\n",
    "                p_het,\n",
    "                model_het.params[1],\n",
    "                model_het.params[2],\n",
    "                model_het.bse[1],\n",
    "                model_het.bse[2],\n",
    "            ]\n",
    "    df_assoc_trait.sort_values([\"CHROM\", \"POS\"]).to_csv(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "019fd055-ed3f-434f-9613-e4d0b4ccb2f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=20,\n",
    "    memory_g=16,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_locus_hetero,\n",
    "    trait_list,\n",
    "    [f\"out/locus-hetero/{trait}.csv\" for trait in trait_list],\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
