{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48f561e5-437c-4833-bdc9-1c8d147a0eb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "import submitit\n",
    "import matplotlib.pyplot as plt\n",
    "import admix_genet_cor\n",
    "import admix\n",
    "from os.path import join\n",
    "import os\n",
    "import glob\n",
    "import scipy\n",
    "\n",
    "SUPP_TABLE_URL = \"https://www.dropbox.com/s/jck2mhjby2ur55j/supp_tables.xlsx?dl=1\"\n",
    "ROOT_DIR = (\n",
    "    \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01-dataset/out/aframr\"\n",
    ")\n",
    "PFILE_DIR = join(ROOT_DIR, \"imputed\")\n",
    "trait_list = [\n",
    "    f.split(\"/\")[-1].split(\".\")[0] for f in glob.glob(join(ROOT_DIR, \"pheno\", \"*.tsv\"))\n",
    "]\n",
    "trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"trait-info\")\n",
    "trait_list = trait_info[\"trait\"].values\n",
    "dict_trait_display_name = {\n",
    "    row[\"trait\"]: row[\"display-name\"] for _, row in trait_info.iterrows()\n",
    "}\n",
    "GRM_DIR = \"/u/scratch/k/kangchen/admix-grm/rho-model\"\n",
    "\n",
    "CLUMP_DIR = (\n",
    "    \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/02-gwas/out/aframr\"\n",
    ")\n",
    "\n",
    "df_params = pd.DataFrame(\n",
    "    [\n",
    "        params\n",
    "        for params in itertools.product(\n",
    "            [\"hm3\", \"imputed\"],\n",
    "            [\"mafukb\", \"gcta\"],\n",
    "            [0.005, 0.05],\n",
    "        )\n",
    "    ],\n",
    "    columns=[\n",
    "        \"snpset\",\n",
    "        \"hermodel\",\n",
    "        \"maf\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "df_params[\"grm_prefix\"] = df_params.apply(\n",
    "    lambda p: f\"{p.snpset}.{p.hermodel}.{str(p.maf)[2:]}\",\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "\n",
    "def submit_gcta_estimate(grm_prefix, trait, duffy_covar=True):\n",
    "    # compile phenotype and covariates\n",
    "    dset = admix.io.read_dataset(\n",
    "        join(PFILE_DIR, \"chr1\"),\n",
    "        n_anc=2,\n",
    "    )\n",
    "    df_trait = pd.read_csv(join(ROOT_DIR, f\"pheno/{trait}.tsv\"), sep=\"\\t\", index_col=0)\n",
    "    df_trait.index = df_trait.index.astype(str)\n",
    "\n",
    "    # subset for individuals with non-nan value in df_trait\n",
    "    dset = dset[:, dset.indiv.index.isin(df_trait.index)]\n",
    "    dset.append_indiv_info(df_trait)\n",
    "\n",
    "    covar_cols = df_trait.columns[1:]\n",
    "\n",
    "    df_pheno = dset.indiv[[trait]].copy()\n",
    "    df_covar = dset.indiv[covar_cols].copy()\n",
    "    df_covar = admix.data.convert_dummy(df_covar)\n",
    "\n",
    "    # special case for duffy SNPs, include the duffy SNPs in the covariate\n",
    "    if duffy_covar:\n",
    "        # find closest SNPs\n",
    "        duffy_snp_loc = np.argmin(np.abs(dset.snp.POS - 159204893))\n",
    "        assert dset.snp.CHROM.iloc[duffy_snp_loc] == 1\n",
    "        duffy_lanc = dset[duffy_snp_loc].lanc.sum(axis=[0, 2]).compute()\n",
    "        df_covar[\"duffy_lanc\"] = duffy_lanc\n",
    "\n",
    "    ### include GWAS variants as covariates\n",
    "    clump_path = join(CLUMP_DIR, trait, \"PLINK.imputed.clumped\")\n",
    "    if (sum(1 for line in open(clump_path)) > 1) and trait != \"total_wbc_cnt\":\n",
    "        # when there is clumped variants\n",
    "        df_clump = pd.read_csv(\n",
    "            join(CLUMP_DIR, trait, \"PLINK.imputed.clumped\"), delim_whitespace=True\n",
    "        )\n",
    "        # include variants with h2 > 0.004\n",
    "        df_clump = df_clump[df_clump.P < scipy.stats.chi2.sf(dset.n_indiv * 0.004, 1)]\n",
    "        if len(df_clump) > 0:\n",
    "            print(f\"{len(df_clump)} SNPs with expected h2 > 0.2%\")\n",
    "            df_clump_geno = []\n",
    "            for chrom, df_chrom_clump in df_clump.groupby(\"CHR\"):\n",
    "                dset_chrom = admix.io.read_dataset(\n",
    "                    join(PFILE_DIR, f\"chr{chrom}\"),\n",
    "                    n_anc=2,\n",
    "                )\n",
    "                dset_chrom = dset_chrom[\n",
    "                    df_chrom_clump.SNP.values, dset.indiv.index.values\n",
    "                ]\n",
    "                df_clump_geno.append(\n",
    "                    pd.DataFrame(\n",
    "                        dset_chrom.geno.sum(axis=2).T.compute(),\n",
    "                        columns=dset_chrom.snp.index.values,\n",
    "                        index=dset_chrom.indiv.index.values,\n",
    "                    )\n",
    "                )\n",
    "            df_clump_geno = pd.concat(df_clump_geno, axis=1)\n",
    "            print(df_clump_geno)\n",
    "            df_covar = pd.merge(\n",
    "                df_covar, df_clump_geno, left_index=True, right_index=True\n",
    "            )\n",
    "\n",
    "    for col in df_pheno.columns:\n",
    "        df_pheno[col] = admix.data.quantile_normalize(df_pheno[col])\n",
    "\n",
    "    for col in df_covar.columns:\n",
    "        df_covar[col] = admix.data.quantile_normalize(df_covar[col])\n",
    "\n",
    "    # fill na with column mean\n",
    "    df_covar.fillna(df_covar.mean(), inplace=True)\n",
    "\n",
    "    df_id = pd.DataFrame(\n",
    "        {\"FID\": df_pheno.index.values, \"IID\": df_pheno.index.values},\n",
    "        index=df_pheno.index.values,\n",
    "    )\n",
    "    df_pheno = pd.merge(df_id, df_pheno, left_index=True, right_index=True)\n",
    "    df_covar = pd.merge(df_id, df_covar, left_index=True, right_index=True)\n",
    "\n",
    "    out_dir = f\"out/gcta-estimate/{trait}-{grm_prefix}\"\n",
    "    os.makedirs(out_dir, exist_ok=True)\n",
    "    rho_list = np.linspace(0, 1, 21)\n",
    "\n",
    "    ### fit different rho\n",
    "    for rho in rho_list:\n",
    "        grm = join(GRM_DIR, f\"{grm_prefix}/rho{int(rho * 100)}\")\n",
    "        out_prefix = os.path.join(out_dir, f\"rho{int(rho * 100)}\")\n",
    "        if not os.path.exists(out_prefix + \".hsq\"):\n",
    "            admix.tools.gcta.reml(\n",
    "                grm_path=grm,\n",
    "                df_pheno=df_pheno,\n",
    "                df_covar=df_covar,\n",
    "                out_prefix=out_prefix,\n",
    "                n_thread=4,\n",
    "            )\n",
    "\n",
    "df_params = pd.DataFrame(\n",
    "    [params for params in itertools.product(df_params.grm_prefix.unique(), trait_list)],\n",
    "    columns=[\"grm_prefix\", \"trait\"],\n",
    ")\n",
    "df_todo_params = df_params[\n",
    "    df_params.apply(\n",
    "        lambda x: not os.path.exists(\n",
    "            f\"out/gcta-estimate/{x.trait}-{x.grm_prefix}/rho100.hsq\"\n",
    "        ),\n",
    "        axis=1,\n",
    "    )\n",
    "]\n",
    "df_todo_params\n",
    "\n",
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=700,\n",
    "    memory_g=20,\n",
    "    #     queue=\"highp\",\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_gcta_estimate,\n",
    "    df_todo_params.grm_prefix,\n",
    "    df_todo_params.trait,\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
