{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e27b8e0c-f4ef-455a-a2c5-118d13289158",
   "metadata": {},
   "source": [
    "# Loci analysis\n",
    "1. Take the top SNPs for each significant loci\n",
    "2. Test the heterogeneity score.\n",
    "3. Aggregate over multiple traits and multiple regions and show QQ-plot\n",
    "\n",
    "`page-gwas.tsv` is downloaded from GWAS catalog https://www.ebi.ac.uk/gwas/publications/31217584 (Download catalog data)\n",
    "\n",
    "**Links below are not used in this notebook but I preserve these just in case**\n",
    "\n",
    "Also see \n",
    "- https://github.com/gokceneraslan/opentargets-genetics-python\n",
    "- https://community.opentargets.org/t/how-to-access-finngen-gwas-data-using-the-open-targets-genetics-portal-api/254/4\n",
    "- https://api.genetics.opentargets.org/graphql/schema\n",
    "\n",
    "!wget https://static-content.springer.com/esm/art%3A10.1038%2Fs41586-019-1310-4/MediaObjects/41586_2019_1310_MOESM3_ESM.xlsx -O page-supp-tables.xlsx\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b9a3a36-9d68-463b-b58a-cb58a5932b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%load_ext lab_black\n",
    "\n",
    "import xarray as xr\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import admix\n",
    "import dask\n",
    "import dask.array as da\n",
    "from tqdm import tqdm\n",
    "import statsmodels.api as sm\n",
    "import matplotlib.pyplot as plt\n",
    "from admix.data import quantile_normalize\n",
    "import submitit\n",
    "\n",
    "import admix_genet_cor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b713c7e7-1fc6-4f7e-9f37-9903688f66e7",
   "metadata": {},
   "source": [
    "# Process GWAS hits from PAGE study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d32e6ce5-efff-4813-85f6-2430f675ae06",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPP_TABLE_URL = \"https://www.dropbox.com/s/jck2mhjby2ur55j/supp_tables.xlsx?dl=1\"\n",
    "trait_info = pd.read_excel(SUPP_TABLE_URL, sheet_name=\"trait-info\")\n",
    "trait_list = trait_info[\"trait\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2f328f0e-0979-4c9a-8dde-6094456d6653",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_assoc = pd.read_csv(\"./page-gwas.tsv\", sep=\"\\t\")\n",
    "df_assoc = (\n",
    "    df_assoc[\n",
    "        [\n",
    "            \"DISEASE/TRAIT\",\n",
    "            \"INITIAL SAMPLE SIZE\",\n",
    "            \"REGION\",\n",
    "            \"SNPS\",\n",
    "            \"CHR_ID\",\n",
    "            \"CHR_POS\",\n",
    "            \"P-VALUE\",\n",
    "            \"STUDY ACCESSION\",\n",
    "        ]\n",
    "    ]\n",
    "    .dropna(subset=[\"CHR_POS\"])\n",
    "    .astype({\"CHR_POS\": int})\n",
    ")\n",
    "\n",
    "df_assoc = df_assoc.loc[df_assoc.CHR_ID.isin(np.arange(1, 23).astype(str))]\n",
    "# NOTE: the 24 + 2 additional traits are waist-hip ratio for males and females\n",
    "\n",
    "# convert trait_id\n",
    "gwas_catalog_name2id = {\n",
    "    row[\"GWAS catalog name\"]: row[\"trait\"] for _, row in trait_info.iterrows()\n",
    "}\n",
    "df_assoc.insert(\n",
    "    0, \"trait_id\", df_assoc[\"DISEASE/TRAIT\"].apply(lambda x: gwas_catalog_name2id[x])\n",
    ")\n",
    "df_assoc[\"CHR_ID\"] = df_assoc[\"CHR_ID\"].astype(int)\n",
    "df_assoc = (\n",
    "    df_assoc.sort_values([\"trait_id\", \"CHR_ID\", \"CHR_POS\"])\n",
    "    .drop_duplicates([\"trait_id\", \"CHR_ID\", \"CHR_POS\"])\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "df_assoc.index = (\n",
    "    df_assoc[\"trait_id\"]\n",
    "    + \":\"\n",
    "    + df_assoc[\"CHR_ID\"].astype(str)\n",
    "    + \":\"\n",
    "    + df_assoc[\"CHR_POS\"].astype(str)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5d6cd2dc-5ca1-4894-be95-39db786f80f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dapgen\n",
    "from os.path import join\n",
    "\n",
    "PFILE_DIR = \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01_dataset/out/aframr/imputed\"\n",
    "SAMPLE_INFO_PATH = \"/u/project/pasaniuc/pasaniucdata/admixture/projects/PAGE-QC/01_dataset/out/aframr/sample_info.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "96447c4c-a6af-471a-8ed7-c7a05a6295a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_locus_hetero(trait, out):\n",
    "    df_assoc_trait = df_assoc[df_assoc.trait_id == trait].copy()\n",
    "\n",
    "    # format phenotype and covariates\n",
    "\n",
    "    dset = admix.dataset.read_dataset(\n",
    "        join(PFILE_DIR, \"chr22\"),\n",
    "        indiv_info=SAMPLE_INFO_PATH,\n",
    "        n_anc=2,\n",
    "    )\n",
    "    dset_assoc = dset[:, ~np.isnan(dset.indiv[trait]).values]\n",
    "    covar_cols = [\"age\", \"sex\", \"study\"] + [f\"geno_EV{i}\" for i in range(1, 11)]\n",
    "\n",
    "    df_pheno = dset_assoc.indiv[[trait]].copy()\n",
    "    df_covar = dset_assoc.indiv[covar_cols].copy()\n",
    "    # create study dummies variables\n",
    "    study_dummies = pd.get_dummies(df_covar[\"study\"], drop_first=True)\n",
    "    study_dummies.columns = [f\"study_dummy_{s}\" for s in study_dummies.columns]\n",
    "    df_covar = pd.concat([df_covar, study_dummies], axis=1)\n",
    "    df_covar = df_covar.drop(columns=[\"study\"])\n",
    "\n",
    "    for col in df_pheno.columns:\n",
    "        df_pheno[col] = admix.data.quantile_normalize(df_pheno[col])\n",
    "\n",
    "    for col in df_covar.columns:\n",
    "        df_covar[col] = admix.data.quantile_normalize(df_covar[col])\n",
    "\n",
    "    for chrom in range(1, 23):\n",
    "        # subset df_assoc\n",
    "        df_assoc_trait_chrom = df_assoc_trait[df_assoc_trait.CHR_ID == chrom]\n",
    "\n",
    "        # subset dset\n",
    "        dset = admix.dataset.read_dataset(\n",
    "            join(PFILE_DIR, f\"chr{chrom}\"),\n",
    "            indiv_info=SAMPLE_INFO_PATH,\n",
    "            n_anc=2,\n",
    "        )\n",
    "        dset = dset[\n",
    "            dset.snp.POS.isin(df_assoc_trait_chrom.CHR_POS.values).values,\n",
    "            df_pheno.index.values,\n",
    "        ]\n",
    "        dset.persist()\n",
    "\n",
    "        apa = dset.allele_per_anc()\n",
    "        af = dset.af_per_anc()\n",
    "\n",
    "        for snp_i in tqdm(range(dset.n_snp)):\n",
    "            p_het, model_het = admix_genet_cor.test_snp_het(\n",
    "                apa[snp_i, :, :], df_pheno.values, df_covar.values\n",
    "            )\n",
    "            p_assoc, model_assoc = admix_genet_cor.test_snp_assoc(\n",
    "                apa[snp_i, :, :], df_pheno.values, df_covar.values\n",
    "            )\n",
    "            snp_idx = f\"{trait}:{dset.snp.CHROM[snp_i]}:{dset.snp.POS[snp_i]}\"\n",
    "            df_assoc_trait.loc[snp_idx, [\"EUR_af\", \"AFR_af\"]] = (\n",
    "                af[snp_i, 0],\n",
    "                af[snp_i, 1],\n",
    "            )\n",
    "\n",
    "            df_assoc_trait.loc[\n",
    "                snp_idx,\n",
    "                [\n",
    "                    \"assoc_pval\",\n",
    "                    \"HET_pval\",\n",
    "                    \"EUR_beta\",\n",
    "                    \"AFR_beta\",\n",
    "                    \"EUR_beta_stderr\",\n",
    "                    \"AFR_beta_stderr\",\n",
    "                ],\n",
    "            ] = [\n",
    "                p_assoc,\n",
    "                p_het,\n",
    "                model_het.params[1],\n",
    "                model_het.params[2],\n",
    "                model_het.bse[1],\n",
    "                model_het.bse[2],\n",
    "            ]\n",
    "    df_assoc_trait.to_csv(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "019fd055-ed3f-434f-9613-e4d0b4ccb2f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "executor = submitit.SgeExecutor(folder=\"./submitit-logs\")\n",
    "\n",
    "executor.update_parameters(\n",
    "    time_min=20,\n",
    "    memory_g=16,\n",
    "    setup=[\n",
    "        \"export PATH=~/project-pasaniuc/software/miniconda3/bin:$PATH\",\n",
    "        \"export PYTHONNOUSERSITE=True\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "jobs = executor.map_array(\n",
    "    submit_locus_hetero,\n",
    "    trait_list,\n",
    "    [f\"out/locus_hetero/{trait}.csv\" for trait in trait_list],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7156033a-b209-4487-8624-016634157c4e",
   "metadata": {},
   "source": [
    "# DEPRECATED BELOW\n",
    "-----------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d24ad4f3-dbce-4e80-b90c-ffb5bb31f702",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for chrom in range(1, 23):\n",
    "#     df_loc = (\n",
    "#         df_assoc[df_assoc[\"CHR_ID\"] == str(chrom)][[\"CHR_ID\", \"CHR_POS\"]]\n",
    "#         .drop_duplicates()\n",
    "#         .sort_values([\"CHR_POS\"])\n",
    "#     )\n",
    "#     pfile = join(PFILE_DIR, f\"chr{chrom}\")\n",
    "#     df_snp = dapgen.read_pvar(pfile + \".pvar\")\n",
    "#     snp_mask = df_snp.POS.isin(df_loc.CHR_POS)\n",
    "\n",
    "#     print(f\"chrom={chrom}, {sum(snp_mask)}/{len(df_loc)} SNPs is found\")\n",
    "\n",
    "#     lanc = admix.io.read_lanc(pfile + \".lanc\")[snp_mask, :, :].compute()\n",
    "#     lanc = da.from_array(lanc)\n",
    "\n",
    "#     admix.tools.plink2_subset(\n",
    "#         pfile, f\"out/gwas_hit_geno/chr{chrom}\", df_snp.index.values[snp_mask]\n",
    "#     )\n",
    "#     admix.io.write_lanc(f\"out/gwas_hit_geno/chr{chrom}.lanc\", lanc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "94eef75f-565b-4c71-8b54-f8f90bab56e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1/1 [02:11<00:00, 131.89s/it]\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p out/locus_het/assoc_loc\n",
    "for chrom in tqdm(range(1, 2)):\n",
    "    df_loc = (\n",
    "        df_assoc[df_assoc[\"CHR_ID\"] == str(chrom)][[\"CHR_ID\", \"CHR_POS\"]]\n",
    "        .drop_duplicates()\n",
    "        .sort_values([\"CHR_POS\"])\n",
    "    )\n",
    "    df_loc[\"CHR_ID\"] = \"chr\" + df_loc[\"CHR_ID\"]\n",
    "    df_loc.rename(columns={\"CHR_ID\": \"CHROM\", \"CHR_POS\": \"POS\"}).to_csv(\n",
    "        f\"out/locus_het/assoc_loc/chr{chrom}.tsv\", sep=\"\\t\", index=False, header=False\n",
    "    )\n",
    "    cmd = (\n",
    "        f\"tabix -hR out/locus_het/assoc_loc/chr{chrom}.tsv \"\n",
    "        + f\"/u/project/sgss/PAGE/ImputedGWAS_topmedfrz8/MEGA_all.chr{chrom}.filtered.vcf.gz \"\n",
    "        + f\"> out/locus_het/assoc_loc/chr{chrom}.vcf\"\n",
    "    )\n",
    "    subprocess.check_call(cmd, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c9a340-0440-46a1-9d53-fff6ef2ed420",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load dset_gwas_hit\n",
    "dset_list = []\n",
    "for chrom in tqdm(range(1, 23)):\n",
    "    dset_list.append(admix.io.read_vcf(f\"out/locus_het/assoc_loc/chr{chrom}.vcf\"))\n",
    "dset_gwas_hit = xr.concat(dset_list, dim=\"snp\")\n",
    "dset_gwas_hit[\"geno\"] = dset_gwas_hit.geno.dims, da.from_array(\n",
    "    dset_gwas_hit.geno.values, chunks=-1\n",
    ")\n",
    "dset_gwas_hit.attrs[\"n_anc\"] = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "3051bc8f-8aaf-4241-8298-d5bd5970e504",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "904it [00:03, 301.23it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<xarray.backends.zarr.ZarrStore at 0x2aaffe7f5cf0>"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# align `dset_gwas_hit to `dset`\n",
    "dset_gwas_hit = dset_gwas_hit.sel(indiv=dset_hm3.indiv.values)\n",
    "\n",
    "# fill in lanc\n",
    "df_dset_hm3 = dset_hm3.snp.to_dataframe().reset_index(drop=True)\n",
    "snp_index = []\n",
    "for _, snp in tqdm(dset_gwas_hit.snp.to_dataframe().iterrows()):\n",
    "    chrom, pos = snp[\"CHROM\"], snp[\"POS\"]\n",
    "    df_tmp = df_dset_hm3[df_dset_hm3[\"CHROM\"] == chrom]\n",
    "    snp_index.append((df_tmp[\"POS\"] - pos).abs().idxmin())\n",
    "lanc = dset_hm3.isel(snp=snp_index).lanc.values\n",
    "dset_gwas_hit = dset_gwas_hit.assign(\n",
    "    lanc=((\"indiv\", \"snp\", \"ploidy\"), da.from_array(lanc, chunks=-1))\n",
    ")\n",
    "\n",
    "# fill in individual information\n",
    "for k in dset_hm3.coords:\n",
    "    if dset_hm3.coords[k].dims == (\"indiv\",):\n",
    "        dset_gwas_hit.coords[k] = (\"indiv\", dset_hm3.coords[k].data)\n",
    "\n",
    "# for duplicated snps (with same CHROM and POS), retain only the one with larger MAF\n",
    "df_tmp = dset_gwas_hit.snp.to_dataframe()\n",
    "df_tmp = df_tmp.groupby([\"CHROM\", \"POS\"], as_index=False).apply(\n",
    "    lambda group: group.loc[group.MAF == group.MAF.max()]\n",
    ")\n",
    "dset_gwas_hit = dset_gwas_hit.sel(snp=dset_gwas_hit.snp.isin(df_tmp.snp))\n",
    "admix.tools.allele_per_anc(dset_gwas_hit)\n",
    "dset_gwas_hit.to_zarr(\"out/locus_het/gwas_hit.zarr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "580a6016-c139-424a-bced-5cb91043ef3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_assoc[\"EUR_af\"] = np.nan\n",
    "df_assoc[\"AFR_af\"] = np.nan\n",
    "\n",
    "df_assoc[\"EUR_beta\"] = np.nan\n",
    "df_assoc[\"EUR_beta_stderr\"] = np.nan\n",
    "df_assoc[\"AFR_beta\"] = np.nan\n",
    "df_assoc[\"AFR_beta_stderr\"] = np.nan\n",
    "df_assoc[\"HET_pval\"] = np.nan\n",
    "\n",
    "for row_i, row in tqdm(df_assoc.iterrows(), total=df_assoc.shape[0]):\n",
    "    chrom = int(row.CHR_ID)\n",
    "    position = row.CHR_POS\n",
    "\n",
    "    dset_snp_pos = np.where(\n",
    "        (dset_gwas_hit.CHROM == chrom) & (dset_gwas_hit.POS == position)\n",
    "    )[0]\n",
    "    if len(dset_snp_pos) == 0:\n",
    "        print(f\"{row_i} is missing\")\n",
    "        continue\n",
    "    assert len(dset_snp_pos) == 1\n",
    "    dset_snp = dset_gwas_hit.isel(snp=dset_snp_pos)\n",
    "    dset_snp = dset_snp.sel(indiv=~np.isnan(dset_snp.coords[row.trait_id]))\n",
    "    apa = dset_snp[\"allele_per_anc\"].values[:, 0, :]\n",
    "    admix.tools.af_per_anc(dset_snp)\n",
    "    df_covar = dict()\n",
    "    for col in [\"age\", \"sex\", \"study\"] + [f\"geno_EV{i}\" for i in range(1, 11)]:\n",
    "        df_covar[col] = dset_snp.coords[col].values\n",
    "    df_covar = pd.DataFrame(df_covar)\n",
    "    study_dummies = pd.get_dummies(df_covar[\"study\"], drop_first=True)\n",
    "    study_dummies.columns = [f\"study_dummy_{s}\" for s in study_dummies.columns]\n",
    "    df_covar = pd.concat([df_covar, study_dummies], axis=1)\n",
    "    df_covar = df_covar.drop(columns=[\"study\"])\n",
    "    covar = df_covar.values\n",
    "    covar = (covar - covar.mean(axis=0)) / covar.std(axis=0)\n",
    "    y = dset_snp[row.trait_id].values\n",
    "    y = quantile_normalize(y)\n",
    "    p_het, model_het = common.test_het(apa, y, covar)\n",
    "    p_assoc, model_assoc = common.test_assoc(apa, y, covar)\n",
    "    df_assoc.loc[row_i, [\"EUR_af\", \"AFR_af\"]] = (\n",
    "        dset_snp[\"af_per_anc\"].values[0, 0],\n",
    "        dset_snp[\"af_per_anc\"].values[0, 1],\n",
    "    )\n",
    "    df_assoc.loc[\n",
    "        row_i,\n",
    "        [\n",
    "            \"assoc_pval\",\n",
    "            \"HET_pval\",\n",
    "            \"EUR_beta\",\n",
    "            \"AFR_beta\",\n",
    "            \"EUR_beta_stderr\",\n",
    "            \"AFR_beta_stderr\",\n",
    "        ],\n",
    "    ] = [\n",
    "        p_assoc,\n",
    "        p_het,\n",
    "        model_het.params[1],\n",
    "        model_het.params[2],\n",
    "        model_het.bse[1],\n",
    "        model_het.bse[2],\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "a9c898b3-24dc-4dab-b371-cecb8782a70d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_assoc.to_excel(\"out/locus_het/sumstats.xlsx\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
